// @fluxstack/live/dist/build/index.js
import { existsSync, rmSync, mkdirSync, writeFileSync, readFileSync } from 'fs';
import { join, resolve, dirname } from 'path';

// src/build/index.ts
var BUILD_VERSION = "0.1.0";
function extractMeta(filePath) {
  const src = readFileSync(filePath, "utf-8");
  const results = [];
  const re = /export\s+class\s+(\w+)\s+extends\s+LiveComponent/g;
  let m;
  while ((m = re.exec(src)) !== null) {
    const className = m[1];
    const body = extractBlock(src, src.indexOf("{", m.index));
    const name = body.match(/static\s+componentName\s*=\s*['"]([^'"]+)['"]/)?.[1] ?? className;
    const actions = body.match(/static\s+publicActions\s*=\s*(\[[^\]]*\])/)?.[1] ?? "[]";
    const state = extractDefaultState(body);
    results.push({ className, componentName: name, defaultState: state, publicActions: actions });
  }
  return results;
}
function extractBlock(src, start) {
  let depth = 1, i = start + 1;
  while (i < src.length && depth > 0) {
    if (src[i] === "{") depth++;
    else if (src[i] === "}") depth--;
    i++;
  }
  return src.substring(start, i);
}
function extractDefaultState(classBody) {
  const m = classBody.match(/static\s+defaultState\s*=\s*/);
  if (!m) return "{}";
  const objStart = classBody.indexOf("{", m.index + m[0].length);
  if (objStart === -1) return "{}";
  const raw = extractBlock(classBody, objStart);
  return stripAsCasts(raw);
}
function stripAsCasts(s) {
  const RE = /\s+as\s+/g;
  let out = "", last = 0, m;
  while ((m = RE.exec(s)) !== null) {
    out += s.slice(last, m.index);
    let i = m.index + m[0].length;
    const stack = [];
    while (i < s.length) {
      const c = s[i];
      if (c === "{" || c === "<" || c === "(") {
        stack.push(c === "{" ? "}" : c === "<" ? ">" : ")");
        i++;
      } else if (c === "[" && s[i + 1] === "]") {
        i += 2;
      } else if (c === "[") {
        stack.push("]");
        i++;
      } else if (stack.length && c === stack[stack.length - 1]) {
        stack.pop();
        i++;
        while (s[i] === "[" && s[i + 1] === "]") i += 2;
      } else if (!stack.length && (c === "," || c === "\n" || c === "}")) break;
      else i++;
    }
    last = i;
  }
  return out + s.slice(last);
}
function buildStub(metas) {
  if (!metas.length) return "export {}";
  return metas.map(
    (m) => `export class ${m.className} {
  static componentName = '${m.componentName}'
  static defaultState = ${m.defaultState}
  static publicActions = ${m.publicActions}
}`
  ).join("\n\n");
}
function norm(p) {
  return p.replace(/\\/g, "/");
}
function liveStripPlugin(options = {}) {
  const {
    importPrefix = "@server/live/",
    stubDir: stubDirName = ".live-stubs",
    verbose = false
  } = options;
  let projectRoot;
  let stubDir;
  const nameToFile = /* @__PURE__ */ new Map();
  const fileToName = /* @__PURE__ */ new Map();
  const cache = /* @__PURE__ */ new Map();
  const log = verbose ? (msg) => console.log(`[live-strip] ${msg}`) : () => {
  };
  function writeStub(name, serverPath) {
    const stubPath = join(stubDir, `${name}.js`);
    const content = buildStub(extractMeta(serverPath));
    if (cache.get(name) !== content) {
      writeFileSync(stubPath, content, "utf-8");
      cache.set(name, content);
      log(`Generated stub: ${name}`);
    }
    return stubPath;
  }
  return {
    name: "fluxstack-live-strip",
    enforce: "pre",
    configResolved(config) {
      projectRoot = config.configFile ? dirname(config.configFile) : resolve(config.root, "../..");
      stubDir = join(config.root, stubDirName);
      if (!existsSync(stubDir)) mkdirSync(stubDir, { recursive: true });
    },
    resolveId(source, importer) {
      if (!source.startsWith(importPrefix) || !importer) return null;
      const imp = norm(importer);
      if (!imp.includes("/client/") && !imp.includes("/app/client/")) return null;
      const name = source.replace(importPrefix, "");
      let serverBase;
      if (options.serverDir) {
        serverBase = resolve(projectRoot, options.serverDir);
      } else {
        serverBase = resolve(projectRoot, source.replace("@server/", "app/server/"));
      }
      const ts = serverBase.endsWith(".ts") ? serverBase : serverBase + ".ts";
      nameToFile.set(name, ts);
      fileToName.set(norm(ts), name);
      return writeStub(name, ts);
    },
    handleHotUpdate({ file, server }) {
      const name = fileToName.get(norm(file));
      if (!name) return;
      const serverPath = nameToFile.get(name);
      const oldContent = cache.get(name);
      const newContent = buildStub(extractMeta(serverPath));
      if (newContent === oldContent) return [];
      writeStub(name, serverPath);
      const stubPath = norm(join(stubDir, `${name}.js`));
      const mods = server.moduleGraph.getModulesByFile(stubPath);
      if (mods?.size) {
        const arr = [...mods];
        arr.forEach((m) => server.moduleGraph.invalidateModule(m));
        server.config.logger.info(`[live-strip] HMR: ${name} metadata changed`, { timestamp: true });
        return arr;
      }
    },
    buildEnd() {
      if (existsSync(stubDir)) rmSync(stubDir, { recursive: true, force: true });
    }
  };
}

export { BUILD_VERSION, liveStripPlugin };
