// @fluxstack/sdk/dist/chunk-JZAQK34Z.js
// env.ts
var EnvLoader = class {
  cache = /* @__PURE__ */ new Map();
  getEnvSource() {
    const g = globalThis;
    const bun = g["Bun"];
    if (bun?.["env"]) return bun["env"];
    const proc = g["process"];
    if (proc?.["env"]) return proc["env"];
    return {};
  }
  get(key, defaultValue) {
    const cacheKey = `${key}:${typeof defaultValue}`;
    if (this.cache.has(cacheKey)) return this.cache.get(cacheKey);
    const source = this.getEnvSource();
    const raw = source[key];
    if (!raw || raw === "") {
      this.cache.set(cacheKey, defaultValue);
      return defaultValue;
    }
    let result = raw;
    if (typeof defaultValue === "number") {
      const n = Number(raw);
      result = isNaN(n) ? defaultValue : n;
    } else if (typeof defaultValue === "boolean") {
      result = ["true", "1", "yes", "on"].includes(raw.toLowerCase());
    } else if (Array.isArray(defaultValue)) {
      result = raw.split(",").map((v) => v.trim()).filter(Boolean);
    } else if (typeof defaultValue === "object" && defaultValue !== null) {
      try {
        result = JSON.parse(raw);
      } catch {
        result = defaultValue;
      }
    }
    this.cache.set(cacheKey, result);
    return result;
  }
  has(key) {
    const source = this.getEnvSource();
    const v = source[key];
    return v !== void 0 && v !== "";
  }
  all() {
    const source = this.getEnvSource();
    const out = {};
    for (const [k, v] of Object.entries(source)) {
      if (v !== void 0 && v !== "") out[k] = v;
    }
    return out;
  }
  clearCache() {
    this.cache.clear();
  }
};
var loader = new EnvLoader();
var env = {
  get: (key, defaultValue) => loader.get(key, defaultValue),
  has: (key) => loader.has(key),
  all: () => loader.all(),
  clearCache: () => loader.clearCache()
};
var env_default = env;

export {
  env,
  env_default
};
