// @fluxstack/sdk/dist/config.js
import {
  env
} from "./chunk-JZAQK34Z.js";

// config.ts
function castValue(value, type) {
  if (value === void 0 || value === null) return void 0;
  switch (type) {
    case "string":
      return String(value);
    case "number": {
      const num = Number(value);
      return isNaN(num) ? void 0 : num;
    }
    case "boolean":
      if (typeof value === "boolean") return value;
      if (typeof value === "string") return ["true", "1", "yes", "on"].includes(value.toLowerCase());
      return Boolean(value);
    case "array":
      if (Array.isArray(value)) return value;
      if (typeof value === "string") return value.split(",").map((v) => v.trim()).filter(Boolean);
      return [value];
    case "object":
      if (typeof value === "object" && value !== null) return value;
      if (typeof value === "string") {
        try {
          return JSON.parse(value);
        } catch {
          return {};
        }
      }
      return {};
    case "enum":
      return value;
    default:
      return value;
  }
}
function validateField(fieldName, value, field) {
  if (field.required && (value === void 0 || value === null || value === "")) {
    return { field: fieldName, message: `Field '${fieldName}' is required but not provided` };
  }
  if (value === void 0 && !field.required) return null;
  if (field.type === "enum" && field.values) {
    if (!field.values.includes(value)) {
      return { field: fieldName, message: `Field '${fieldName}' must be one of: ${field.values.join(", ")}`, value };
    }
  }
  if (field.validate) {
    const result = field.validate(value);
    if (result === false) return { field: fieldName, message: `Field '${fieldName}' failed validation`, value };
    if (typeof result === "string") return { field: fieldName, message: result, value };
  }
  return null;
}
function defineConfig(schema) {
  const result = {};
  const errors = [];
  for (const [fieldName, field] of Object.entries(schema)) {
    let value;
    if (field.env) {
      const envValue = env.has(field.env) ? env.all()[field.env] : void 0;
      if (envValue !== void 0 && envValue !== "") value = envValue;
    }
    if (value === void 0) value = field.default;
    if (value !== void 0 && field.transform) {
      try {
        value = field.transform(value);
      } catch (err) {
        errors.push({ field: fieldName, message: `Transform failed: ${err}` });
        continue;
      }
    }
    if (value !== void 0) value = castValue(value, field.type);
    const error = validateField(fieldName, value, field);
    if (error) {
      errors.push(error);
      continue;
    }
    result[fieldName] = value;
  }
  if (errors.length > 0) {
    const msg = errors.map((e) => `  - ${e.message}${e.value !== void 0 ? ` (got: ${JSON.stringify(e.value)})` : ""}`).join("\n");
    throw new Error(`Configuration validation failed:
${msg}`);
  }
  return result;
}
function envString(envVar, defaultValue, required = false) {
  return { type: "string", env: envVar, default: defaultValue, required };
}
function envNumber(envVar, defaultValue, required = false) {
  return { type: "number", env: envVar, default: defaultValue, required };
}
function envBoolean(envVar, defaultValue, required = false) {
  return { type: "boolean", env: envVar, default: defaultValue, required };
}
function envArray(envVar, defaultValue, required = false) {
  return { type: "array", env: envVar, default: defaultValue, required };
}
function envEnum(envVar, values, defaultValue, required = false) {
  return { type: "enum", env: envVar, values, default: defaultValue, required };
}
var config = {
  string: envString,
  number: envNumber,
  boolean: envBoolean,
  array: envArray,
  enum: envEnum
};
export {
  config,
  defineConfig,
  envArray,
  envBoolean,
  envEnum,
  envNumber,
  envString
};
