// @fluxstack/sdk/config.ts
/**
 * @fluxstack/plugin-sdk/config
 *
 * Declarative configuration system for FluxStack plugins.
 * Self-contained — no dependency on FluxStack internals.
 *
 * @example
 * ```ts
 * import { defineConfig, config } from '@fluxstack/plugin-sdk/config'
 *
 * const myConfig = defineConfig({
 *   enabled: config.boolean('MY_PLUGIN_ENABLED', true),
 *   port: config.number('MY_PLUGIN_PORT', 8080),
 *   mode: config.enum('MY_PLUGIN_MODE', ['fast', 'safe'] as const, 'safe'),
 * } as const)
 *
 * myConfig.enabled  // boolean
 * myConfig.port     // number
 * myConfig.mode     // "fast" | "safe"
 * ```
 */

import { env } from './env'

// ─── Field types ─────────────────────────────────────────────

export type ConfigFieldType = 'string' | 'number' | 'boolean' | 'array' | 'object' | 'enum'

export interface ConfigField<T = unknown> {
  type: ConfigFieldType
  env?: string
  default?: T
  required?: boolean
  validate?(value: T): boolean | string
  values?: readonly T[]
  description?: string
  transform?: (value: unknown) => T
}

export type ConfigSchema = Record<string, ConfigField>

// ─── Type inference ──────────────────────────────────────────

export type InferConfig<T extends ConfigSchema> = {
  [K in keyof T]: T[K]['default'] extends infer D
    ? D extends undefined
      ? T[K]['required'] extends true
        ? InferFieldType<T[K]>
        : InferFieldType<T[K]> | undefined
      : InferFieldType<T[K]>
    : InferFieldType<T[K]>
}

type InferFieldType<F> =
  F extends ConfigField<infer T>
    ? T extends undefined
      ? (
        F extends { type: 'string' } ? string :
        F extends { type: 'number' } ? number :
        F extends { type: 'boolean' } ? boolean :
        F extends { type: 'array' } ? string[] :
        F extends { type: 'object' } ? Record<string, unknown> :
        F extends { type: 'enum'; values: readonly (infer U)[] } ? U :
        unknown
      )
      : T
    : unknown

// ─── Validation ──────────────────────────────────────────────

export interface ValidationError {
  field: string
  message: string
  value?: unknown
}

export interface ValidationResult {
  valid: boolean
  errors: ValidationError[]
  warnings?: string[]
}

function castValue(value: unknown, type: ConfigFieldType): unknown {
  if (value === undefined || value === null) return undefined

  switch (type) {
    case 'string':
      return String(value)
    case 'number': {
      const num = Number(value)
      return isNaN(num) ? undefined : num
    }
    case 'boolean':
      if (typeof value === 'boolean') return value
      if (typeof value === 'string') return ['true', '1', 'yes', 'on'].includes(value.toLowerCase())
      return Boolean(value)
    case 'array':
      if (Array.isArray(value)) return value
      if (typeof value === 'string') return value.split(',').map(v => v.trim()).filter(Boolean)
      return [value]
    case 'object':
      if (typeof value === 'object' && value !== null) return value
      if (typeof value === 'string') { try { return JSON.parse(value) } catch { return {} } }
      return {}
    case 'enum':
      return value
    default:
      return value
  }
}

function validateField(
  fieldName: string,
  value: unknown,
  field: ConfigField
): ValidationError | null {
  if (field.required && (value === undefined || value === null || value === '')) {
    return { field: fieldName, message: `Field '${fieldName}' is required but not provided` }
  }
  if (value === undefined && !field.required) return null
  if (field.type === 'enum' && field.values) {
    if (!field.values.includes(value)) {
      return { field: fieldName, message: `Field '${fieldName}' must be one of: ${field.values.join(', ')}`, value }
    }
  }
  if (field.validate) {
    const result = field.validate(value)
    if (result === false) return { field: fieldName, message: `Field '${fieldName}' failed validation`, value }
    if (typeof result === 'string') return { field: fieldName, message: result, value }
  }
  return null
}

// ─── defineConfig ────────────────────────────────────────────

export function defineConfig<T extends ConfigSchema>(schema: T): InferConfig<T> {
  const result: Record<string, unknown> = {}
  const errors: ValidationError[] = []

  for (const [fieldName, field] of Object.entries(schema)) {
    let value: unknown

    // 1. Read from env
    if (field.env) {
      const envValue = env.has(field.env) ? env.all()[field.env] : undefined
      if (envValue !== undefined && envValue !== '') value = envValue
    }

    // 2. Default
    if (value === undefined) value = field.default

    // 3. Transform
    if (value !== undefined && field.transform) {
      try { value = field.transform(value) } catch (err) {
        errors.push({ field: fieldName, message: `Transform failed: ${err}` })
        continue
      }
    }

    // 4. Cast
    if (value !== undefined) value = castValue(value, field.type)

    // 5. Validate
    const error = validateField(fieldName, value, field)
    if (error) { errors.push(error); continue }

    result[fieldName] = value
  }

  if (errors.length > 0) {
    const msg = errors
      .map(e => `  - ${e.message}${e.value !== undefined ? ` (got: ${JSON.stringify(e.value)})` : ''}`)
      .join('\n')
    throw new Error(`Configuration validation failed:\n${msg}`)
  }

  return result as InferConfig<T>
}

// ─── Shorthand helpers ───────────────────────────────────────

export function envString(envVar: string, defaultValue?: string, required = false): ConfigField<string> {
  return { type: 'string', env: envVar, default: defaultValue, required }
}

export function envNumber(envVar: string, defaultValue?: number, required = false): ConfigField<number> {
  return { type: 'number', env: envVar, default: defaultValue, required }
}

export function envBoolean(envVar: string, defaultValue?: boolean, required = false): ConfigField<boolean> {
  return { type: 'boolean', env: envVar, default: defaultValue, required }
}

export function envArray(envVar: string, defaultValue?: string[], required = false): ConfigField<string[]> {
  return { type: 'array', env: envVar, default: defaultValue, required }
}

export function envEnum<T extends readonly string[]>(
  envVar: string,
  values: T,
  defaultValue?: T[number],
  required = false
): ConfigField<T[number]> {
  return { type: 'enum', env: envVar, values, default: defaultValue, required }
}

export const config = {
  string: envString,
  number: envNumber,
  boolean: envBoolean,
  array: envArray,
  enum: envEnum,
}
