// @fluxstack/sdk/index.ts
/**
 * @fluxstack/plugin-sdk
 *
 * Public types for building FluxStack plugins.
 * Import types from this package instead of @core/plugins/types.
 *
 * @example
 * ```ts
 * import type { Plugin, PluginContext, ValidationContext } from '@fluxstack/plugin-sdk'
 * ```
 */

// ─── Hook types ──────────────────────────────────────────────

export type PluginHook =
  // Lifecycle hooks
  | 'setup'
  | 'onConfigLoad'
  | 'onBeforeServerStart'
  | 'onServerStart'
  | 'onAfterServerStart'
  | 'onBeforeServerStop'
  | 'onServerStop'
  // Request/Response pipeline hooks
  | 'onRequest'
  | 'onBeforeRoute'
  | 'onAfterRoute'
  | 'onBeforeResponse'
  | 'onResponse'
  | 'onRequestValidation'
  | 'onResponseTransform'
  // Error handling hooks
  | 'onError'
  // Build pipeline hooks
  | 'onBeforeBuild'
  | 'onBuild'
  | 'onBuildAsset'
  | 'onBuildComplete'
  | 'onBuildError'
  // Plugin system hooks
  | 'onPluginRegister'
  | 'onPluginUnregister'
  | 'onPluginError'

export type PluginPriority = 'highest' | 'high' | 'normal' | 'low' | 'lowest' | number

// ─── Context types ───────────────────────────────────────────

export interface PluginContext {
  config: Record<string, unknown>
  logger: PluginLogger
  app: unknown
  utils: PluginUtils
  registry?: unknown
}

export interface PluginLogger {
  debug: (...args: unknown[]) => void
  info: (...args: unknown[]) => void
  warn: (...args: unknown[]) => void
  error: (...args: unknown[]) => void
}

export interface PluginUtils {
  createTimer: (label: string) => { end: () => number }
  formatBytes: (bytes: number) => string
  isProduction: () => boolean
  isDevelopment: () => boolean
  getEnvironment: () => string
  createHash: (data: string) => string
  deepMerge: (target: Record<string, unknown>, source: Record<string, unknown>) => Record<string, unknown>
  validateSchema: (data: Record<string, unknown>, schema: PluginConfigSchema) => { valid: boolean; errors: string[] }
}

export interface RequestContext {
  request: Request
  path: string
  method: string
  headers: Record<string, string>
  query: Record<string, string>
  params: Record<string, string>
  body?: unknown
  user?: unknown
  startTime: number
  handled?: boolean
  response?: Response
}

export interface ResponseContext extends RequestContext {
  response: Response
  statusCode: number
  duration: number
  size?: number
}

export interface ErrorContext extends RequestContext {
  error: Error
  duration: number
  handled: boolean
}

export interface BuildContext {
  target: string
  outDir: string
  mode: 'development' | 'production'
  config: Record<string, unknown>
}

export interface ConfigLoadContext {
  config: Record<string, unknown>
  envVars: Record<string, string | undefined>
  configPath?: string
}

export interface RouteContext extends RequestContext {
  route?: string
  handler?: Function
  params: Record<string, string>
}

export interface ValidationContext extends RequestContext {
  errors: Array<{ field: string; message: string; code: string }>
  isValid: boolean
}

export interface TransformContext extends ResponseContext {
  transformed: boolean
  originalResponse?: Response
}

export interface BuildAssetContext {
  assetPath: string
  assetType: 'js' | 'css' | 'html' | 'image' | 'font' | 'other'
  size: number
  content?: string | Uint8Array
}

export interface BuildErrorContext {
  error: Error
  file?: string
  line?: number
  column?: number
}

export interface PluginEventContext {
  pluginName: string
  pluginVersion?: string
  timestamp: number
  data?: unknown
}

export interface PluginConfigSchema {
  type: 'object'
  properties: Record<string, unknown>
  required?: string[]
  additionalProperties?: boolean
}

// ─── Plugin interface ────────────────────────────────────────

export interface Plugin {
  name: string
  version?: string
  description?: string
  author?: string
  dependencies?: string[]
  priority?: number | PluginPriority
  category?: string
  tags?: string[]

  // Lifecycle hooks
  setup?: (context: PluginContext) => void | Promise<void>
  onConfigLoad?: (context: ConfigLoadContext) => void | Promise<void>
  onBeforeServerStart?: (context: PluginContext) => void | Promise<void>
  onServerStart?: (context: PluginContext) => void | Promise<void>
  onAfterServerStart?: (context: PluginContext) => void | Promise<void>
  onBeforeServerStop?: (context: PluginContext) => void | Promise<void>
  onServerStop?: (context: PluginContext) => void | Promise<void>

  // Request/Response pipeline hooks
  onRequest?: (context: RequestContext) => void | Promise<void>
  onBeforeRoute?: (context: RequestContext) => void | Promise<void>
  onAfterRoute?: (context: RouteContext) => void | Promise<void>
  onBeforeResponse?: (context: ResponseContext) => void | Promise<void>
  onResponse?: (context: ResponseContext) => void | Promise<void>
  onRequestValidation?: (context: ValidationContext) => void | Promise<void>
  onResponseTransform?: (context: TransformContext) => void | Promise<void>

  // Error handling hooks
  onError?: (context: ErrorContext) => void | Promise<void>

  // Build pipeline hooks
  onBeforeBuild?: (context: BuildContext) => void | Promise<void>
  onBuild?: (context: BuildContext) => void | Promise<void>
  onBuildAsset?: (context: BuildAssetContext) => void | Promise<void>
  onBuildComplete?: (context: BuildContext) => void | Promise<void>
  onBuildError?: (context: BuildErrorContext) => void | Promise<void>

  // Plugin system hooks
  onPluginRegister?: (context: PluginEventContext) => void | Promise<void>
  onPluginUnregister?: (context: PluginEventContext) => void | Promise<void>
  onPluginError?: (context: PluginEventContext & { error: Error }) => void | Promise<void>

  // Configuration (deprecated — use defineConfig from @fluxstack/plugin-sdk/config)
  configSchema?: PluginConfigSchema
  defaultConfig?: unknown

  // CLI commands
  commands?: CliCommand[]

  // Elysia sub-app (mounted automatically by the framework)
  plugin?: unknown

  // Allow additional properties for extensibility
  [key: string]: unknown
}

// ─── CLI types ───────────────────────────────────────────────

export interface CliArgument {
  name: string
  description: string
  required?: boolean
  type?: 'string' | 'number' | 'boolean'
  default?: unknown
  choices?: string[]
}

export interface CliOption {
  name: string
  short?: string
  description: string
  type?: 'string' | 'number' | 'boolean' | 'array'
  default?: unknown
  required?: boolean
  choices?: string[]
}

export interface CliCommand {
  name: string
  description: string
  usage?: string
  examples?: string[]
  arguments?: CliArgument[]
  options?: CliOption[]
  aliases?: string[]
  category?: string
  hidden?: boolean
  handler: (args: unknown[], options: Record<string, unknown>, context: CliContext) => Promise<void> | void
}

export interface CliContext {
  config: Record<string, unknown>
  logger: PluginLogger
  utils: PluginUtils
  workingDir: string
  packageInfo: {
    name: string
    version: string
  }
}

// ─── Manifest types ──────────────────────────────────────────

export interface PluginManifest {
  name: string
  version: string
  description: string
  author: string
  license: string
  homepage?: string
  repository?: string
  keywords: string[]
  dependencies: Record<string, string>
  peerDependencies?: Record<string, string>
  fluxstack: {
    version: string
    hooks: PluginHook[]
    config?: PluginConfigSchema
    category?: string
    tags?: string[]
    client?: string
  }
}

// ─── Namespace for backward compat ───────────────────────────

export type FluxStackPlugin = Plugin
