// @fluxstack/live/dist/index.js
import { randomBytes, createHmac, createCipheriv, createDecipheriv, scryptSync } from 'crypto';
import { gzipSync, gunzipSync } from 'zlib';
import { EventEmitter } from 'events';

var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/rooms/RoomEventBus.ts
function createTypedRoomEventBus() {
  const subscriptions = /* @__PURE__ */ new Map();
  const componentIndex = /* @__PURE__ */ new Map();
  const getKey = (roomType, roomId, event) => `${roomType}:${roomId}:${event}`;
  const getRoomKey = (roomType, roomId) => `${roomType}:${roomId}`;
  return {
    on(roomType, roomId, event, componentId, handler) {
      const key = getKey(roomType, roomId, event);
      if (!subscriptions.has(key)) {
        subscriptions.set(key, /* @__PURE__ */ new Set());
      }
      const subscription = {
        roomType,
        roomId,
        event,
        handler,
        componentId
      };
      subscriptions.get(key).add(subscription);
      if (!componentIndex.has(componentId)) {
        componentIndex.set(componentId, /* @__PURE__ */ new Set());
      }
      componentIndex.get(componentId).add(key);
      return () => {
        subscriptions.get(key)?.delete(subscription);
        if (subscriptions.get(key)?.size === 0) {
          subscriptions.delete(key);
        }
      };
    },
    emit(roomType, roomId, event, data, excludeComponentId) {
      const key = getKey(roomType, roomId, event);
      const subs = subscriptions.get(key);
      if (!subs || subs.size === 0) return 0;
      let notified = 0;
      for (const sub of subs) {
        if (excludeComponentId && sub.componentId === excludeComponentId) continue;
        try {
          sub.handler(data);
          notified++;
        } catch (error) {
          console.error(`[RoomEventBus] Error in handler [${key}]:`, error);
        }
      }
      return notified;
    },
    unsubscribeAll(componentId) {
      const keys = componentIndex.get(componentId);
      if (!keys) return 0;
      let removed = 0;
      for (const key of keys) {
        const subs = subscriptions.get(key);
        if (!subs) continue;
        for (const sub of subs) {
          if (sub.componentId === componentId) {
            subs.delete(sub);
            removed++;
          }
        }
        if (subs.size === 0) {
          subscriptions.delete(key);
        }
      }
      componentIndex.delete(componentId);
      return removed;
    },
    clearRoom(roomType, roomId) {
      const prefix = getRoomKey(roomType, roomId);
      let removed = 0;
      for (const key of subscriptions.keys()) {
        if (key.startsWith(prefix)) {
          const subs = subscriptions.get(key);
          if (subs) {
            for (const sub of subs) {
              componentIndex.get(sub.componentId)?.delete(key);
            }
            removed += subs.size;
          }
          subscriptions.delete(key);
        }
      }
      return removed;
    },
    getStats() {
      const rooms = {};
      let total = 0;
      for (const [key, subs] of subscriptions) {
        const parts = key.split(":");
        const roomKey = `${parts[0]}:${parts[1]}`;
        const event = parts[2] ?? "";
        if (!rooms[roomKey]) {
          rooms[roomKey] = { events: {} };
        }
        rooms[roomKey].events[event] = subs.size;
        total += subs.size;
      }
      return { totalSubscriptions: total, rooms };
    }
  };
}
var RoomEventBus = class {
  subscriptions = /* @__PURE__ */ new Map();
  /** Reverse index: componentId -> Set of subscription keys for O(1) unsubscribeAll */
  componentIndex = /* @__PURE__ */ new Map();
  getKey(roomType, roomId, event) {
    return `${roomType}:${roomId}:${event}`;
  }
  on(roomType, roomId, event, componentId, handler) {
    const key = this.getKey(roomType, roomId, event);
    if (!this.subscriptions.has(key)) {
      this.subscriptions.set(key, /* @__PURE__ */ new Set());
    }
    const subscription = { roomType, roomId, event, handler, componentId };
    this.subscriptions.get(key).add(subscription);
    if (!this.componentIndex.has(componentId)) {
      this.componentIndex.set(componentId, /* @__PURE__ */ new Set());
    }
    this.componentIndex.get(componentId).add(key);
    return () => {
      this.subscriptions.get(key)?.delete(subscription);
      if (this.subscriptions.get(key)?.size === 0) {
        this.subscriptions.delete(key);
      }
    };
  }
  emit(roomType, roomId, event, data, excludeComponentId) {
    const key = this.getKey(roomType, roomId, event);
    const subs = this.subscriptions.get(key);
    if (!subs || subs.size === 0) return 0;
    let notified = 0;
    for (const sub of subs) {
      if (excludeComponentId && sub.componentId === excludeComponentId) continue;
      try {
        sub.handler(data);
        notified++;
      } catch (error) {
        console.error(`[RoomEventBus] Error in handler [${key}]:`, error);
      }
    }
    return notified;
  }
  unsubscribeAll(componentId) {
    const keys = this.componentIndex.get(componentId);
    if (!keys) return 0;
    let removed = 0;
    for (const key of keys) {
      const subs = this.subscriptions.get(key);
      if (!subs) continue;
      for (const sub of subs) {
        if (sub.componentId === componentId) {
          subs.delete(sub);
          removed++;
        }
      }
      if (subs.size === 0) {
        this.subscriptions.delete(key);
      }
    }
    this.componentIndex.delete(componentId);
    return removed;
  }
  clearRoom(roomType, roomId) {
    const prefix = `${roomType}:${roomId}`;
    let removed = 0;
    for (const key of this.subscriptions.keys()) {
      if (key.startsWith(prefix)) {
        const subs = this.subscriptions.get(key);
        if (subs) {
          for (const sub of subs) {
            this.componentIndex.get(sub.componentId)?.delete(key);
          }
          removed += subs.size;
        }
        this.subscriptions.delete(key);
      }
    }
    return removed;
  }
  getStats() {
    const rooms = {};
    let total = 0;
    for (const [key, subs] of this.subscriptions) {
      const parts = key.split(":");
      const roomKey = `${parts[0]}:${parts[1]}`;
      const event = parts[2] ?? "";
      if (!rooms[roomKey]) {
        rooms[roomKey] = { events: {} };
      }
      rooms[roomKey].events[event] = subs.size;
      total += subs.size;
    }
    return { totalSubscriptions: total, rooms };
  }
};

// src/protocol/constants.ts
var PROTOCOL_VERSION = 1;
var DEFAULT_WS_PATH = "/api/live/ws";
var DEFAULT_CHUNK_SIZE = 64 * 1024;
var DEFAULT_RATE_LIMIT_MAX_TOKENS = 100;
var DEFAULT_RATE_LIMIT_REFILL_RATE = 50;
var MAX_MESSAGE_SIZE = 100 * 1024 * 1024;
var MAX_ROOM_STATE_SIZE = 10 * 1024 * 1024;
var MAX_ROOMS_PER_CONNECTION = 50;
var ROOM_NAME_REGEX = /^[a-zA-Z0-9_:.-]{1,64}$/;
var MAX_QUEUE_SIZE = 1e3;

// src/transport/WsSendBatcher.ts
var wsQueues = /* @__PURE__ */ new WeakMap();
var scheduledFlushes = /* @__PURE__ */ new WeakSet();
var pendingWs = [];
var globalFlushScheduled = false;
function scheduleWs(ws) {
  if (!scheduledFlushes.has(ws)) {
    scheduledFlushes.add(ws);
    pendingWs.push(ws);
    if (!globalFlushScheduled) {
      globalFlushScheduled = true;
      queueMicrotask(flushAll);
    }
  }
}
function queueWsMessage(ws, message) {
  if (!ws || ws.readyState !== 1) return;
  let queue = wsQueues.get(ws);
  if (!queue) {
    queue = [];
    wsQueues.set(ws, queue);
  }
  if (queue.length >= MAX_QUEUE_SIZE) {
    queue.shift();
  }
  queue.push(message);
  scheduleWs(ws);
}
function queuePreSerialized(ws, serialized) {
  if (!ws || ws.readyState !== 1) return;
  let queue = wsQueues.get(ws);
  if (!queue) {
    queue = [];
    wsQueues.set(ws, queue);
  }
  if (queue.length >= MAX_QUEUE_SIZE) {
    queue.shift();
  }
  queue.push(serialized);
  scheduleWs(ws);
}
function flushAll() {
  globalFlushScheduled = false;
  const connections = pendingWs;
  pendingWs = [];
  for (const ws of connections) {
    scheduledFlushes.delete(ws);
    const queue = wsQueues.get(ws);
    if (!queue || queue.length === 0) continue;
    wsQueues.set(ws, []);
    if (ws.readyState !== 1) continue;
    try {
      if (queue.length === 1) {
        const item = queue[0];
        if (typeof item === "string") {
          ws.send(item);
        } else {
          ws.send(JSON.stringify(item));
        }
      } else {
        const objects = [];
        const preSerialized = [];
        for (const item of queue) {
          if (typeof item === "string") {
            preSerialized.push(item);
          } else {
            objects.push(item);
          }
        }
        if (objects.length === 0) {
          ws.send("[" + preSerialized.join(",") + "]");
        } else if (preSerialized.length === 0) {
          const deduped = deduplicateDeltas(objects);
          ws.send(JSON.stringify(deduped));
        } else {
          const deduped = deduplicateDeltas(objects);
          let result = "[";
          for (let i = 0; i < deduped.length; i++) {
            if (i > 0) result += ",";
            result += JSON.stringify(deduped[i]);
          }
          for (const ps of preSerialized) {
            result += "," + ps;
          }
          result += "]";
          ws.send(result);
        }
      }
    } catch {
    }
  }
}
function deduplicateDeltas(messages) {
  const deltaIndices = /* @__PURE__ */ new Map();
  const result = [];
  for (const msg of messages) {
    if (msg.type === "STATE_DELTA" && msg.componentId && msg.payload?.delta) {
      const existing = deltaIndices.get(msg.componentId);
      if (existing !== void 0) {
        const target = result[existing];
        target.payload = {
          delta: { ...target.payload.delta, ...msg.payload.delta }
        };
        target.timestamp = msg.timestamp;
      } else {
        deltaIndices.set(msg.componentId, result.length);
        result.push({ ...msg, payload: { delta: { ...msg.payload.delta } } });
      }
    } else {
      result.push(msg);
    }
  }
  return result;
}
function sendBinaryImmediate(ws, data) {
  if (ws && ws.readyState === 1) {
    ws.send(data);
  }
}
function sendImmediate(ws, data) {
  if (ws && ws.readyState === 1) {
    ws.send(data);
  }
}

// src/debug/LiveLogger.ts
var componentConfigs = /* @__PURE__ */ new Map();
var globalConfigParsed = false;
var globalConfig = false;
function parseGlobalConfig() {
  if (globalConfigParsed) return globalConfig;
  globalConfigParsed = true;
  const envValue = typeof process !== "undefined" ? process.env?.LIVE_LOGGING : void 0;
  if (!envValue || envValue === "false") {
    globalConfig = false;
  } else if (envValue === "true") {
    globalConfig = true;
  } else {
    globalConfig = envValue.split(",").map((s) => s.trim()).filter(Boolean);
  }
  return globalConfig;
}
function registerComponentLogging(componentId, config) {
  if (config !== void 0 && config !== false) {
    componentConfigs.set(componentId, config);
  }
}
function unregisterComponentLogging(componentId) {
  componentConfigs.delete(componentId);
}
function shouldLog(componentId, category) {
  if (componentId) {
    const config = componentConfigs.get(componentId);
    if (config === void 0 || config === false) return false;
    if (config === true) return true;
    return config.includes(category);
  }
  const cfg = parseGlobalConfig();
  if (cfg === false) return false;
  if (cfg === true) return true;
  return cfg.includes(category);
}
function liveLog(category, componentId, message, ...args) {
  if (shouldLog(componentId, category)) {
    if (args.length > 0) {
      console.log(message, ...args);
    } else {
      console.log(message);
    }
  }
}
function liveWarn(category, componentId, message, ...args) {
  if (shouldLog(componentId, category)) {
    if (args.length > 0) {
      console.warn(message, ...args);
    } else {
      console.warn(message);
    }
  }
}

// src/utils/deepDiff.ts
function isPlainObject(v) {
  return v !== null && typeof v === "object" && !Array.isArray(v) && Object.getPrototypeOf(v) === Object.prototype;
}
function computeDeepDiff(prev, next, depth = 0, maxDepth = 3, seen) {
  if (depth > maxDepth) return prev === next ? null : next;
  if (!seen) seen = /* @__PURE__ */ new Set();
  if (seen.has(next)) return prev === next ? null : next;
  seen.add(next);
  let result = null;
  for (const key of Object.keys(next)) {
    const oldVal = prev[key];
    const newVal = next[key];
    if (oldVal === newVal) continue;
    if (isPlainObject(oldVal) && isPlainObject(newVal)) {
      const nested = computeDeepDiff(oldVal, newVal, depth + 1, maxDepth, seen);
      if (nested !== null) {
        result ??= {};
        result[key] = nested;
      }
    } else {
      result ??= {};
      result[key] = newVal;
    }
  }
  return result;
}
function deepAssign(target, source, seen) {
  if (!seen) seen = /* @__PURE__ */ new Set();
  if (seen.has(source)) return;
  seen.add(source);
  for (const key of Object.keys(source)) {
    if (isPlainObject(target[key]) && isPlainObject(source[key])) {
      deepAssign(target[key], source[key], seen);
    } else {
      target[key] = source[key];
    }
  }
}

// src/rooms/RoomCodec.ts
var BINARY_ROOM_EVENT = 2;
var BINARY_ROOM_STATE = 3;
var encoder = new TextEncoder();
var decoder = new TextDecoder();
function msgpackEncode(value) {
  const parts = [];
  encode(value, parts);
  let totalLen = 0;
  for (const p of parts) totalLen += p.length;
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const p of parts) {
    result.set(p, offset);
    offset += p.length;
  }
  return result;
}
function encode(value, parts) {
  if (value === null || value === void 0) {
    parts.push(new Uint8Array([192]));
    return;
  }
  if (typeof value === "boolean") {
    parts.push(new Uint8Array([value ? 195 : 194]));
    return;
  }
  if (typeof value === "number") {
    if (Number.isInteger(value)) {
      encodeInt(value, parts);
    } else {
      const buf = new Uint8Array(9);
      buf[0] = 203;
      new DataView(buf.buffer).setFloat64(1, value, false);
      parts.push(buf);
    }
    return;
  }
  if (typeof value === "string") {
    const encoded = encoder.encode(value);
    const len = encoded.length;
    if (len < 32) {
      parts.push(new Uint8Array([160 | len]));
    } else if (len < 256) {
      parts.push(new Uint8Array([217, len]));
    } else if (len < 65536) {
      parts.push(new Uint8Array([218, len >> 8, len & 255]));
    } else {
      parts.push(new Uint8Array([219, len >> 24 & 255, len >> 16 & 255, len >> 8 & 255, len & 255]));
    }
    parts.push(encoded);
    return;
  }
  if (value instanceof Uint8Array) {
    const len = value.length;
    if (len < 256) {
      parts.push(new Uint8Array([196, len]));
    } else if (len < 65536) {
      parts.push(new Uint8Array([197, len >> 8, len & 255]));
    } else {
      parts.push(new Uint8Array([198, len >> 24 & 255, len >> 16 & 255, len >> 8 & 255, len & 255]));
    }
    parts.push(value);
    return;
  }
  if (Array.isArray(value)) {
    const len = value.length;
    if (len < 16) {
      parts.push(new Uint8Array([144 | len]));
    } else if (len < 65536) {
      parts.push(new Uint8Array([220, len >> 8, len & 255]));
    } else {
      parts.push(new Uint8Array([221, len >> 24 & 255, len >> 16 & 255, len >> 8 & 255, len & 255]));
    }
    for (const item of value) {
      encode(item, parts);
    }
    return;
  }
  if (typeof value === "object") {
    const keys = Object.keys(value);
    const len = keys.length;
    if (len < 16) {
      parts.push(new Uint8Array([128 | len]));
    } else if (len < 65536) {
      parts.push(new Uint8Array([222, len >> 8, len & 255]));
    } else {
      parts.push(new Uint8Array([223, len >> 24 & 255, len >> 16 & 255, len >> 8 & 255, len & 255]));
    }
    for (const key of keys) {
      encode(key, parts);
      encode(value[key], parts);
    }
    return;
  }
  parts.push(new Uint8Array([192]));
}
function encodeInt(value, parts) {
  if (value >= 0) {
    if (value < 128) {
      parts.push(new Uint8Array([value]));
    } else if (value < 256) {
      parts.push(new Uint8Array([204, value]));
    } else if (value < 65536) {
      parts.push(new Uint8Array([205, value >> 8, value & 255]));
    } else if (value < 4294967296) {
      parts.push(new Uint8Array([206, value >> 24 & 255, value >> 16 & 255, value >> 8 & 255, value & 255]));
    } else {
      const buf = new Uint8Array(9);
      buf[0] = 203;
      new DataView(buf.buffer).setFloat64(1, value, false);
      parts.push(buf);
    }
  } else {
    if (value >= -32) {
      parts.push(new Uint8Array([value & 255]));
    } else if (value >= -128) {
      parts.push(new Uint8Array([208, value & 255]));
    } else if (value >= -32768) {
      const buf = new Uint8Array(3);
      buf[0] = 209;
      new DataView(buf.buffer).setInt16(1, value, false);
      parts.push(buf);
    } else if (value >= -2147483648) {
      const buf = new Uint8Array(5);
      buf[0] = 210;
      new DataView(buf.buffer).setInt32(1, value, false);
      parts.push(buf);
    } else {
      const buf = new Uint8Array(9);
      buf[0] = 203;
      new DataView(buf.buffer).setFloat64(1, value, false);
      parts.push(buf);
    }
  }
}
function msgpackDecode(buf) {
  const result = decodeAt(buf, 0);
  return result.value;
}
function decodeAt(buf, offset) {
  if (offset >= buf.length) return { value: null, offset };
  const byte = buf[offset];
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  if (byte < 128) return { value: byte, offset: offset + 1 };
  if (byte >= 128 && byte <= 143) return decodeMap(buf, offset + 1, byte & 15);
  if (byte >= 144 && byte <= 159) return decodeArray(buf, offset + 1, byte & 15);
  if (byte >= 160 && byte <= 191) {
    const len = byte & 31;
    const str = decoder.decode(buf.subarray(offset + 1, offset + 1 + len));
    return { value: str, offset: offset + 1 + len };
  }
  if (byte >= 224) return { value: byte - 256, offset: offset + 1 };
  switch (byte) {
    // nil
    case 192:
      return { value: null, offset: offset + 1 };
    // false
    case 194:
      return { value: false, offset: offset + 1 };
    // true
    case 195:
      return { value: true, offset: offset + 1 };
    // bin 8
    case 196: {
      const len = buf[offset + 1];
      return { value: buf.slice(offset + 2, offset + 2 + len), offset: offset + 2 + len };
    }
    // bin 16
    case 197: {
      const len = view.getUint16(offset + 1, false);
      return { value: buf.slice(offset + 3, offset + 3 + len), offset: offset + 3 + len };
    }
    // bin 32
    case 198: {
      const len = view.getUint32(offset + 1, false);
      return { value: buf.slice(offset + 5, offset + 5 + len), offset: offset + 5 + len };
    }
    // float 64
    case 203:
      return { value: view.getFloat64(offset + 1, false), offset: offset + 9 };
    // uint 8
    case 204:
      return { value: buf[offset + 1], offset: offset + 2 };
    // uint 16
    case 205:
      return { value: view.getUint16(offset + 1, false), offset: offset + 3 };
    // uint 32
    case 206:
      return { value: view.getUint32(offset + 1, false), offset: offset + 5 };
    // int 8
    case 208:
      return { value: view.getInt8(offset + 1), offset: offset + 2 };
    // int 16
    case 209:
      return { value: view.getInt16(offset + 1, false), offset: offset + 3 };
    // int 32
    case 210:
      return { value: view.getInt32(offset + 1, false), offset: offset + 5 };
    // str 8
    case 217: {
      const len = buf[offset + 1];
      return { value: decoder.decode(buf.subarray(offset + 2, offset + 2 + len)), offset: offset + 2 + len };
    }
    // str 16
    case 218: {
      const len = view.getUint16(offset + 1, false);
      return { value: decoder.decode(buf.subarray(offset + 3, offset + 3 + len)), offset: offset + 3 + len };
    }
    // str 32
    case 219: {
      const len = view.getUint32(offset + 1, false);
      return { value: decoder.decode(buf.subarray(offset + 5, offset + 5 + len)), offset: offset + 5 + len };
    }
    // array 16
    case 220:
      return decodeArray(buf, offset + 3, view.getUint16(offset + 1, false));
    // array 32
    case 221:
      return decodeArray(buf, offset + 5, view.getUint32(offset + 1, false));
    // map 16
    case 222:
      return decodeMap(buf, offset + 3, view.getUint16(offset + 1, false));
    // map 32
    case 223:
      return decodeMap(buf, offset + 5, view.getUint32(offset + 1, false));
  }
  return { value: null, offset: offset + 1 };
}
function decodeArray(buf, offset, count) {
  const arr = new Array(count);
  for (let i = 0; i < count; i++) {
    const result = decodeAt(buf, offset);
    arr[i] = result.value;
    offset = result.offset;
  }
  return { value: arr, offset };
}
function decodeMap(buf, offset, count) {
  const obj = {};
  for (let i = 0; i < count; i++) {
    const keyResult = decodeAt(buf, offset);
    offset = keyResult.offset;
    const valResult = decodeAt(buf, offset);
    offset = valResult.offset;
    obj[String(keyResult.value)] = valResult.value;
  }
  return { value: obj, offset };
}
var msgpackCodec = {
  encode: msgpackEncode,
  decode: msgpackDecode
};
var jsonCodec = {
  encode(data) {
    return encoder.encode(JSON.stringify(data));
  },
  decode(buf) {
    return JSON.parse(decoder.decode(buf));
  }
};
function resolveCodec(option) {
  if (!option || option === "msgpack") return msgpackCodec;
  if (option === "json") return jsonCodec;
  return option;
}
var textEncoder = encoder;
function buildRoomFrame(frameType, componentId, roomId, event, payload) {
  const compIdBytes = textEncoder.encode(componentId);
  const roomIdBytes = textEncoder.encode(roomId);
  const eventBytes = textEncoder.encode(event);
  const totalLen = 1 + 1 + compIdBytes.length + 1 + roomIdBytes.length + 2 + eventBytes.length + payload.length;
  const frame = new Uint8Array(totalLen);
  let offset = 0;
  frame[offset++] = frameType;
  frame[offset++] = compIdBytes.length;
  frame.set(compIdBytes, offset);
  offset += compIdBytes.length;
  frame[offset++] = roomIdBytes.length;
  frame.set(roomIdBytes, offset);
  offset += roomIdBytes.length;
  frame[offset++] = eventBytes.length >> 8 & 255;
  frame[offset++] = eventBytes.length & 255;
  frame.set(eventBytes, offset);
  offset += eventBytes.length;
  frame.set(payload, offset);
  return frame;
}
function buildRoomFrameTail(roomId, event, payload) {
  const roomIdBytes = textEncoder.encode(roomId);
  const eventBytes = textEncoder.encode(event);
  const tailLen = 1 + roomIdBytes.length + 2 + eventBytes.length + payload.length;
  const tail = new Uint8Array(tailLen);
  let offset = 0;
  tail[offset++] = roomIdBytes.length;
  tail.set(roomIdBytes, offset);
  offset += roomIdBytes.length;
  tail[offset++] = eventBytes.length >> 8 & 255;
  tail[offset++] = eventBytes.length & 255;
  tail.set(eventBytes, offset);
  offset += eventBytes.length;
  tail.set(payload, offset);
  return tail;
}
function prependMemberHeader(frameType, componentId, tail) {
  const compIdBytes = textEncoder.encode(componentId);
  const frame = new Uint8Array(1 + 1 + compIdBytes.length + tail.length);
  frame[0] = frameType;
  frame[1] = compIdBytes.length;
  frame.set(compIdBytes, 2);
  frame.set(tail, 2 + compIdBytes.length);
  return frame;
}
function parseRoomFrame(buf) {
  if (buf.length < 6) return null;
  let offset = 0;
  const frameType = buf[offset++];
  const compIdLen = buf[offset++];
  if (offset + compIdLen > buf.length) return null;
  const componentId = decoder.decode(buf.subarray(offset, offset + compIdLen));
  offset += compIdLen;
  const roomIdLen = buf[offset++];
  if (offset + roomIdLen > buf.length) return null;
  const roomId = decoder.decode(buf.subarray(offset, offset + roomIdLen));
  offset += roomIdLen;
  if (offset + 2 > buf.length) return null;
  const eventLen = buf[offset] << 8 | buf[offset + 1];
  offset += 2;
  if (offset + eventLen > buf.length) return null;
  const event = decoder.decode(buf.subarray(offset, offset + eventLen));
  offset += eventLen;
  const payload = buf.subarray(offset);
  return { frameType, componentId, roomId, event, payload };
}

// src/rooms/LiveRoomManager.ts
var LiveRoomManager = class {
  /**
   * @param roomEvents - Local server-side event bus
   * @param pubsub - Optional cross-instance pub/sub adapter (e.g. Redis).
   *                 When provided, room events/state/membership are propagated
   *                 to other server instances in the background.
   */
  constructor(roomEvents, pubsub) {
    this.roomEvents = roomEvents;
    this.pubsub = pubsub;
  }
  rooms = /* @__PURE__ */ new Map();
  componentRooms = /* @__PURE__ */ new Map();
  // componentId -> roomIds
  /** Room registry for LiveRoom class lookup. Set by LiveServer. */
  roomRegistry;
  /**
   * Component joins a room.
   * @param options.deepDiff - Enable/disable deep diff for this room's state. Default: true
   * @param joinContext - Optional context for LiveRoom lifecycle hooks (userId, payload)
   */
  joinRoom(componentId, roomId, ws, initialState, options, joinContext) {
    if (!roomId || !ROOM_NAME_REGEX.test(roomId)) {
      throw new Error("Invalid room name. Must be 1-64 alphanumeric characters, hyphens, underscores, dots, or colons.");
    }
    const now = Date.now();
    let room = this.rooms.get(roomId);
    let isNewRoom = false;
    if (!room) {
      isNewRoom = true;
      const roomClass = this.roomRegistry?.resolveFromId(roomId);
      if (roomClass) {
        const instance = new roomClass(roomId, this);
        const opts = roomClass.$options ?? {};
        room = {
          id: roomId,
          state: instance.state,
          members: /* @__PURE__ */ new Map(),
          createdAt: now,
          lastActivity: now,
          deepDiff: opts.deepDiff ?? true,
          deepDiffDepth: opts.deepDiffDepth ?? 3,
          serverOnlyState: true,
          // LiveRoom-backed rooms are always server-only
          instance,
          codec: resolveCodec(opts.codec)
        };
      } else {
        room = {
          id: roomId,
          state: initialState || {},
          members: /* @__PURE__ */ new Map(),
          createdAt: now,
          lastActivity: now,
          deepDiff: options?.deepDiff ?? true,
          deepDiffDepth: options?.deepDiffDepth ?? 3,
          serverOnlyState: options?.serverOnlyState ?? false,
          codec: null
        };
      }
      this.rooms.set(roomId, room);
      liveLog("rooms", componentId, `Room '${roomId}' created`);
    }
    if (room.instance) {
      const ctor = room.instance.constructor;
      const maxMembers = ctor.$options?.maxMembers;
      if (maxMembers && room.members.size >= maxMembers) {
        return { rejected: true, reason: "Room is full" };
      }
    }
    if (room.instance) {
      const result = room.instance.onJoin({
        componentId,
        userId: joinContext?.userId,
        payload: joinContext?.payload
      });
      if (result === false) {
        if (isNewRoom) {
          this.rooms.delete(roomId);
        }
        return { rejected: true, reason: "Join rejected by room" };
      }
    }
    if (isNewRoom && room.instance) {
      room.instance.onCreate();
    }
    room.members.set(componentId, {
      componentId,
      ws,
      joinedAt: now
    });
    room.lastActivity = now;
    let compRooms = this.componentRooms.get(componentId);
    if (!compRooms) {
      compRooms = /* @__PURE__ */ new Set();
      this.componentRooms.set(componentId, compRooms);
    }
    compRooms.add(roomId);
    const memberCount = room.members.size;
    liveLog("rooms", componentId, `Component '${componentId}' joined room '${roomId}' (${memberCount} members)`);
    this.broadcastToRoom(roomId, {
      type: "ROOM_SYSTEM",
      componentId,
      roomId,
      event: "$sub:join",
      data: {
        subscriberId: componentId,
        count: memberCount
      },
      timestamp: now
    }, componentId);
    this.pubsub?.publishMembership(roomId, "join", componentId)?.catch(() => {
    });
    return { state: room.state };
  }
  /**
   * Component leaves a room
   * @param leaveReason - Why the component is leaving. Default: 'leave'
   */
  leaveRoom(componentId, roomId, leaveReason = "leave") {
    const room = this.rooms.get(roomId);
    if (!room) return;
    if (room.instance) {
      room.instance.onLeave({
        componentId,
        reason: leaveReason
      });
    }
    room.members.delete(componentId);
    const now = Date.now();
    room.lastActivity = now;
    this.componentRooms.get(componentId)?.delete(roomId);
    const memberCount = room.members.size;
    liveLog("rooms", componentId, `Component '${componentId}' left room '${roomId}' (${memberCount} members)`);
    this.broadcastToRoom(roomId, {
      type: "ROOM_SYSTEM",
      componentId,
      roomId,
      event: "$sub:leave",
      data: {
        subscriberId: componentId,
        count: memberCount
      },
      timestamp: now
    });
    this.pubsub?.publishMembership(roomId, "leave", componentId)?.catch(() => {
    });
    if (memberCount === 0) {
      setTimeout(() => {
        const currentRoom = this.rooms.get(roomId);
        if (currentRoom && currentRoom.members.size === 0) {
          if (currentRoom.instance) {
            const result = currentRoom.instance.onDestroy();
            if (result === false) return;
          }
          this.rooms.delete(roomId);
          liveLog("rooms", null, `Room '${roomId}' destroyed (empty)`);
        }
      }, 5 * 60 * 1e3);
    }
  }
  /**
   * Component disconnects - leave all rooms.
   * Batches removals: calls onLeave hooks, removes member from all rooms,
   * then sends leave notifications in bulk.
   */
  cleanupComponent(componentId) {
    const roomIds = this.componentRooms.get(componentId);
    if (!roomIds || roomIds.size === 0) return;
    const now = Date.now();
    const notifications = [];
    for (const roomId of roomIds) {
      const room = this.rooms.get(roomId);
      if (!room) continue;
      if (room.instance) {
        room.instance.onLeave({
          componentId,
          reason: "disconnect"
        });
      }
      room.members.delete(componentId);
      room.lastActivity = now;
      const memberCount = room.members.size;
      if (memberCount > 0) {
        notifications.push({ roomId, count: memberCount });
      } else {
        setTimeout(() => {
          const currentRoom = this.rooms.get(roomId);
          if (currentRoom && currentRoom.members.size === 0) {
            if (currentRoom.instance) {
              const result = currentRoom.instance.onDestroy();
              if (result === false) return;
            }
            this.rooms.delete(roomId);
            liveLog("rooms", null, `Room '${roomId}' destroyed (empty)`);
          }
        }, 5 * 60 * 1e3);
      }
    }
    for (const { roomId, count } of notifications) {
      this.broadcastToRoom(roomId, {
        type: "ROOM_SYSTEM",
        componentId,
        roomId,
        event: "$sub:leave",
        data: {
          subscriberId: componentId,
          count
        },
        timestamp: now
      });
    }
    this.componentRooms.delete(componentId);
  }
  /**
   * Emit event to all members in a room.
   * For LiveRoom-backed rooms, calls onEvent() hook before broadcasting.
   */
  emitToRoom(roomId, event, data, excludeComponentId) {
    const room = this.rooms.get(roomId);
    if (!room) return 0;
    const now = Date.now();
    room.lastActivity = now;
    if (room.instance) {
      room.instance.onEvent(event, data, {
        componentId: excludeComponentId ?? ""
      });
    }
    this.roomEvents.emit("room", roomId, event, data, excludeComponentId);
    this.pubsub?.publish(roomId, event, data)?.catch(() => {
    });
    return this.broadcastToRoom(roomId, {
      type: "ROOM_EVENT",
      componentId: "",
      roomId,
      event,
      data,
      timestamp: now
    }, excludeComponentId);
  }
  /**
   * Update room state.
   * When deepDiff is enabled (default), deep-diffs plain objects to send only changed fields.
   * When disabled, uses shallow diff (reference equality) like classic behavior.
   */
  setRoomState(roomId, updates, excludeComponentId) {
    const room = this.rooms.get(roomId);
    if (!room) return;
    let actualChanges;
    if (room.deepDiff) {
      const diff = computeDeepDiff(
        room.state,
        updates,
        0,
        room.deepDiffDepth
      );
      if (diff === null) return;
      actualChanges = diff;
      deepAssign(room.state, actualChanges);
    } else {
      actualChanges = {};
      let hasChanges = false;
      for (const key of Object.keys(updates)) {
        if (room.state[key] !== updates[key]) {
          actualChanges[key] = updates[key];
          hasChanges = true;
        }
      }
      if (!hasChanges) return;
      Object.assign(room.state, actualChanges);
    }
    if (room.stateSize === void 0) {
      const fullJson = JSON.stringify(room.state);
      room.stateSize = fullJson.length;
      if (room.stateSize > MAX_ROOM_STATE_SIZE) {
        throw new Error("Room state exceeds maximum size limit");
      }
    } else {
      const deltaSize = JSON.stringify(actualChanges).length;
      room.stateSize += deltaSize;
      if (room.stateSize > MAX_ROOM_STATE_SIZE) {
        const precise = JSON.stringify(room.state).length;
        room.stateSize = precise;
        if (precise > MAX_ROOM_STATE_SIZE) {
          throw new Error("Room state exceeds maximum size limit");
        }
      }
    }
    const now = Date.now();
    room.lastActivity = now;
    this.pubsub?.publishStateChange(roomId, actualChanges)?.catch(() => {
    });
    this.broadcastToRoom(roomId, {
      type: "ROOM_STATE",
      componentId: "",
      roomId,
      event: "$state:update",
      data: { state: actualChanges },
      timestamp: now
    }, excludeComponentId);
  }
  /**
   * Get room state
   */
  getRoomState(roomId) {
    return this.rooms.get(roomId)?.state || {};
  }
  /**
   * Broadcast to all members in a room.
   *
   * When the room has a binary codec (LiveRoom-backed), builds a binary frame
   * once (encode payload + frame tail), then prepends per-member componentId header.
   *
   * When no codec (legacy rooms), uses JSON with serialize-once optimization:
   * builds the JSON string template once, then inserts each member's componentId.
   */
  broadcastToRoom(roomId, message, excludeComponentId) {
    const room = this.rooms.get(roomId);
    if (!room || room.members.size === 0) return 0;
    let sent = 0;
    if (room.codec) {
      const frameType = message.type === "ROOM_EVENT" || message.type === "ROOM_SYSTEM" ? BINARY_ROOM_EVENT : BINARY_ROOM_STATE;
      const event = message.event ?? "";
      const payload = room.codec.encode(message.data);
      const tail = buildRoomFrameTail(roomId, event, payload);
      for (const [memberComponentId, member] of room.members) {
        if (memberComponentId === excludeComponentId) continue;
        if (member.ws.readyState !== 1) continue;
        const frame = prependMemberHeader(frameType, memberComponentId, tail);
        sendBinaryImmediate(member.ws, frame);
        sent++;
      }
    } else {
      const { componentId: _, ...rest } = message;
      const jsonBody = JSON.stringify(rest);
      const prefix = '{"componentId":"';
      const suffix = '",' + jsonBody.slice(1);
      for (const [memberComponentId, member] of room.members) {
        if (memberComponentId === excludeComponentId) continue;
        if (member.ws.readyState !== 1) continue;
        queuePreSerialized(member.ws, prefix + memberComponentId + suffix);
        sent++;
      }
    }
    return sent;
  }
  /**
   * Check if component is in a room
   */
  isInRoom(componentId, roomId) {
    return this.rooms.get(roomId)?.members.has(componentId) ?? false;
  }
  /**
   * Check if room state is server-only (no client writes)
   */
  isServerOnlyState(roomId) {
    return this.rooms.get(roomId)?.serverOnlyState ?? false;
  }
  /**
   * Get rooms for a component
   */
  getComponentRooms(componentId) {
    return Array.from(this.componentRooms.get(componentId) || []);
  }
  /**
   * Get member count for a room
   */
  getMemberCount(roomId) {
    return this.rooms.get(roomId)?.members.size ?? 0;
  }
  /**
   * Get the LiveRoom instance for a room (if backed by a typed room class).
   * Used by ComponentRoomProxy to expose custom methods.
   */
  getRoomInstance(roomId) {
    return this.rooms.get(roomId)?.instance;
  }
  /**
   * Get statistics
   */
  getStats() {
    const rooms = {};
    for (const [id, room] of this.rooms) {
      rooms[id] = {
        members: room.members.size,
        createdAt: room.createdAt,
        lastActivity: room.lastActivity
      };
    }
    return {
      totalRooms: this.rooms.size,
      rooms
    };
  }
};

// src/auth/LiveAuthContext.ts
var AuthenticatedContext = class {
  authenticated = true;
  user;
  token;
  authenticatedAt;
  constructor(user, token) {
    this.user = user;
    this.token = token;
    this.authenticatedAt = Date.now();
  }
  hasRole(role) {
    return this.user.roles?.includes(role) ?? false;
  }
  hasAnyRole(roles) {
    if (!this.user.roles?.length) return false;
    return roles.some((role) => this.user.roles.includes(role));
  }
  hasAllRoles(roles) {
    if (!this.user.roles?.length) return roles.length === 0;
    return roles.every((role) => this.user.roles.includes(role));
  }
  hasPermission(permission) {
    return this.user.permissions?.includes(permission) ?? false;
  }
  hasAllPermissions(permissions) {
    if (!this.user.permissions?.length) return permissions.length === 0;
    return permissions.every((perm) => this.user.permissions.includes(perm));
  }
  hasAnyPermission(permissions) {
    if (!this.user.permissions?.length) return false;
    return permissions.some((perm) => this.user.permissions.includes(perm));
  }
};
var AnonymousContext = class {
  authenticated = false;
  user = void 0;
  token = void 0;
  authenticatedAt = void 0;
  hasRole() {
    return false;
  }
  hasAnyRole() {
    return false;
  }
  hasAllRoles() {
    return false;
  }
  hasPermission() {
    return false;
  }
  hasAllPermissions() {
    return false;
  }
  hasAnyPermission() {
    return false;
  }
};
var ANONYMOUS_CONTEXT = new AnonymousContext();

// src/auth/LiveAuthManager.ts
var LiveAuthManager = class {
  providers = /* @__PURE__ */ new Map();
  defaultProviderName;
  /**
   * Register an auth provider.
   */
  register(provider) {
    this.providers.set(provider.name, provider);
    if (!this.defaultProviderName) {
      this.defaultProviderName = provider.name;
    }
    console.log(`[Auth] Provider registered: ${provider.name}`);
  }
  /**
   * Remove an auth provider.
   */
  unregister(name) {
    this.providers.delete(name);
    if (this.defaultProviderName === name) {
      this.defaultProviderName = this.providers.keys().next().value;
    }
  }
  /**
   * Set the default auth provider.
   */
  setDefault(name) {
    if (!this.providers.has(name)) {
      throw new Error(`Auth provider '${name}' not registered`);
    }
    this.defaultProviderName = name;
  }
  /**
   * Returns true if at least one provider is registered.
   */
  hasProviders() {
    return this.providers.size > 0;
  }
  /**
   * Returns the default provider or undefined.
   */
  getDefaultProvider() {
    if (!this.defaultProviderName) return void 0;
    return this.providers.get(this.defaultProviderName);
  }
  /**
   * Authenticate credentials using the specified provider, or try all providers.
   * Returns ANONYMOUS_CONTEXT if no credentials or no providers.
   */
  async authenticate(credentials, providerName) {
    if (!credentials || Object.keys(credentials).every((k) => !credentials[k])) {
      return ANONYMOUS_CONTEXT;
    }
    if (this.providers.size === 0) {
      return ANONYMOUS_CONTEXT;
    }
    if (providerName) {
      const provider = this.providers.get(providerName);
      if (!provider) {
        console.warn(`[Auth] Provider '${providerName}' not found`);
        return ANONYMOUS_CONTEXT;
      }
      try {
        const context = await provider.authenticate(credentials);
        return context || ANONYMOUS_CONTEXT;
      } catch (error) {
        console.error(`[Auth] Failed via '${providerName}':`, error.message);
        return ANONYMOUS_CONTEXT;
      }
    }
    const providersToTry = [];
    if (this.defaultProviderName) {
      const defaultProvider = this.providers.get(this.defaultProviderName);
      if (defaultProvider) providersToTry.push(defaultProvider);
    }
    for (const [name, provider] of this.providers) {
      if (name !== this.defaultProviderName) {
        providersToTry.push(provider);
      }
    }
    const errors = [];
    for (const provider of providersToTry) {
      try {
        const context = await provider.authenticate(credentials);
        if (context && context.authenticated) {
          return context;
        }
      } catch (error) {
        console.warn(`[Auth] Provider '${provider.name}' threw during authentication:`, error.message);
        errors.push({ provider: provider.name, error });
      }
    }
    if (errors.length > 0) {
      console.warn(`[Auth] All ${providersToTry.length} provider(s) failed. Errors: ${errors.map((e) => `${e.provider}: ${e.error.message}`).join("; ")}`);
    }
    return ANONYMOUS_CONTEXT;
  }
  /**
   * Verify auth context meets component requirements.
   */
  authorizeComponent(authContext, authConfig) {
    if (!authConfig) {
      return { allowed: true };
    }
    if (authConfig.required && !authContext.authenticated) {
      return { allowed: false, reason: "Authentication required" };
    }
    if (authConfig.roles?.length) {
      if (!authContext.authenticated) {
        return { allowed: false, reason: `Authentication required. Roles needed: ${authConfig.roles.join(", ")}` };
      }
      if (!authContext.hasAnyRole(authConfig.roles)) {
        return { allowed: false, reason: `Insufficient roles. Required one of: ${authConfig.roles.join(", ")}` };
      }
    }
    if (authConfig.permissions?.length) {
      if (!authContext.authenticated) {
        return { allowed: false, reason: `Authentication required. Permissions needed: ${authConfig.permissions.join(", ")}` };
      }
      if (!authContext.hasAllPermissions(authConfig.permissions)) {
        return { allowed: false, reason: `Insufficient permissions. Required all: ${authConfig.permissions.join(", ")}` };
      }
    }
    return { allowed: true };
  }
  /**
   * Verify auth context allows executing a specific action.
   */
  async authorizeAction(authContext, componentName, action, actionAuth, providerName) {
    if (!actionAuth) {
      return { allowed: true };
    }
    if (actionAuth.roles?.length) {
      if (!authContext.authenticated) {
        return { allowed: false, reason: `Authentication required for action '${action}'` };
      }
      if (!authContext.hasAnyRole(actionAuth.roles)) {
        return { allowed: false, reason: `Insufficient roles for action '${action}'. Required one of: ${actionAuth.roles.join(", ")}` };
      }
    }
    if (actionAuth.permissions?.length) {
      if (!authContext.authenticated) {
        return { allowed: false, reason: `Authentication required for action '${action}'` };
      }
      if (!authContext.hasAllPermissions(actionAuth.permissions)) {
        return { allowed: false, reason: `Insufficient permissions for action '${action}'. Required all: ${actionAuth.permissions.join(", ")}` };
      }
    }
    const name = providerName || this.defaultProviderName;
    if (name) {
      const provider = this.providers.get(name);
      if (provider?.authorizeAction) {
        const allowed = await provider.authorizeAction(authContext, componentName, action);
        if (!allowed) {
          return { allowed: false, reason: `Action '${action}' denied by auth provider '${name}'` };
        }
      }
    }
    return { allowed: true };
  }
  /**
   * Verify auth context allows joining a room.
   */
  async authorizeRoom(authContext, roomId, providerName) {
    const name = providerName || this.defaultProviderName;
    if (!name) return { allowed: true };
    const provider = this.providers.get(name);
    if (!provider?.authorizeRoom) return { allowed: true };
    try {
      const allowed = await provider.authorizeRoom(authContext, roomId);
      if (!allowed) {
        return { allowed: false, reason: `Access to room '${roomId}' denied by auth provider '${name}'` };
      }
      return { allowed: true };
    } catch (error) {
      return { allowed: false, reason: `Room authorization error: ${error.message}` };
    }
  }
  /**
   * Get info about registered providers.
   */
  getInfo() {
    return {
      providers: Array.from(this.providers.keys()),
      defaultProvider: this.defaultProviderName
    };
  }
};
var StateSignatureManager = class {
  secret;
  previousSecrets = [];
  rotationTimer;
  stateBackups = /* @__PURE__ */ new Map();
  config;
  encryptionSalt;
  cachedEncryptionKey = null;
  /** Replay detection: nonce → timestamp when it was first seen. Cleaned every 60s. */
  usedNonces = /* @__PURE__ */ new Map();
  nonceCleanupTimer;
  constructor(config = {}) {
    const defaultSecret = typeof process !== "undefined" ? process.env?.LIVE_STATE_SECRET : void 0;
    this.config = {
      secret: config.secret ?? defaultSecret ?? "",
      rotationEnabled: config.rotationEnabled ?? false,
      rotationInterval: config.rotationInterval ?? 24 * 60 * 60 * 1e3,
      compressionEnabled: config.compressionEnabled ?? true,
      encryptionEnabled: config.encryptionEnabled ?? false,
      nonceEnabled: config.nonceEnabled ?? false,
      maxStateAge: config.maxStateAge ?? 30 * 60 * 1e3,
      backupEnabled: config.backupEnabled ?? true,
      maxBackups: config.maxBackups ?? 3,
      nonceTTL: config.nonceTTL ?? 5 * 60 * 1e3
    };
    if (!this.config.secret) {
      this.config.secret = randomBytes(32).toString("hex");
      liveWarn("state", null, "No LIVE_STATE_SECRET provided. Using random key (state will not persist across restarts).");
    }
    this.secret = Buffer.from(this.config.secret, "utf-8");
    this.encryptionSalt = randomBytes(16);
    if (this.config.rotationEnabled) {
      this.setupKeyRotation();
    }
    if (this.config.nonceEnabled) {
      this.nonceCleanupTimer = setInterval(() => this.cleanupNonces(), this.config.nonceTTL + 10 * 1e3);
    }
  }
  /**
   * Generate a hybrid nonce: `timestamp:random:HMAC(timestamp:random, secret)`
   * Self-validating via HMAC, unique via random bytes, replay-tracked via Map.
   */
  generateNonce() {
    const ts = Date.now().toString();
    const rand = randomBytes(8).toString("hex");
    const payload = `${ts}:${rand}`;
    const mac = createHmac("sha256", this.secret).update(payload).digest("hex").slice(0, 16);
    return `${ts}:${rand}:${mac}`;
  }
  /**
   * Validate a hybrid nonce: check format, HMAC, and TTL.
   */
  validateNonce(nonce) {
    const parts = nonce.split(":");
    if (parts.length !== 3) return { valid: false, error: "Malformed nonce" };
    const [ts, rand, mac] = parts;
    const timestamp = Number(ts);
    if (isNaN(timestamp)) return { valid: false, error: "Malformed nonce timestamp" };
    const age = Date.now() - timestamp;
    if (age > this.config.nonceTTL) {
      return { valid: false, error: "Nonce expired" };
    }
    if (age < -3e4) {
      return { valid: false, error: "Nonce timestamp in the future" };
    }
    const payload = `${ts}:${rand}`;
    const expectedMac = createHmac("sha256", this.secret).update(payload).digest("hex").slice(0, 16);
    if (this.timingSafeEqual(mac, expectedMac)) {
      return { valid: true };
    }
    for (const prevSecret of this.previousSecrets) {
      const prevMac = createHmac("sha256", prevSecret).update(payload).digest("hex").slice(0, 16);
      if (this.timingSafeEqual(mac, prevMac)) {
        return { valid: true };
      }
    }
    return { valid: false, error: "Invalid nonce signature" };
  }
  signState(componentId, state, version, options) {
    let dataStr = JSON.stringify(state);
    let compressed = false;
    let encrypted = false;
    if ((options?.compress ?? this.config.compressionEnabled) && dataStr.length > 1024) {
      const compressedBuf = gzipSync(Buffer.from(dataStr, "utf-8"));
      const compressedB64 = compressedBuf.toString("base64");
      if (compressedB64.length < dataStr.length * 0.9) {
        dataStr = compressedB64;
        compressed = true;
      }
    }
    if (this.config.encryptionEnabled) {
      const iv = randomBytes(16);
      const key = this.deriveEncryptionKey();
      const cipher = createCipheriv("aes-256-cbc", key, iv);
      let encryptedData = cipher.update(dataStr, "utf-8", "base64");
      encryptedData += cipher.final("base64");
      dataStr = iv.toString("base64") + ":" + encryptedData;
      encrypted = true;
    }
    const nonce = this.config.nonceEnabled ? this.generateNonce() : void 0;
    const signedState = {
      data: dataStr,
      signature: "",
      timestamp: Date.now(),
      version,
      componentId,
      nonce,
      compressed,
      encrypted
    };
    signedState.signature = this.computeSignature(signedState);
    if (options?.backup ?? this.config.backupEnabled) {
      this.backupState(componentId, signedState);
    }
    return signedState;
  }
  validateState(signedState, options) {
    try {
      const age = Date.now() - signedState.timestamp;
      if (age > this.config.maxStateAge) {
        return { valid: false, error: "State expired" };
      }
      if (signedState.nonce && this.config.nonceEnabled && !options?.skipNonce) {
        const nonceResult = this.validateNonce(signedState.nonce);
        if (!nonceResult.valid) {
          return { valid: false, error: nonceResult.error };
        }
        if (this.usedNonces.has(signedState.nonce)) {
          return { valid: false, error: "Nonce already used" };
        }
      }
      const expectedSig = this.computeSignature(signedState);
      if (this.timingSafeEqual(signedState.signature, expectedSig)) {
        if (signedState.nonce && this.config.nonceEnabled) {
          this.usedNonces.set(signedState.nonce, Date.now());
        }
        return { valid: true };
      }
      for (const prevSecret of this.previousSecrets) {
        const prevSig = this.computeSignatureWithKey(signedState, prevSecret);
        if (this.timingSafeEqual(signedState.signature, prevSig)) {
          if (signedState.nonce && this.config.nonceEnabled) {
            this.usedNonces.set(signedState.nonce, Date.now());
          }
          return { valid: true };
        }
      }
      return { valid: false, error: "Invalid signature" };
    } catch (error) {
      return { valid: false, error: error.message };
    }
  }
  extractData(signedState) {
    let dataStr = signedState.data;
    if (signedState.encrypted) {
      const [ivB64, encryptedData] = dataStr.split(":");
      const iv = Buffer.from(ivB64, "base64");
      const key = this.deriveEncryptionKey();
      const decipher = createDecipheriv("aes-256-cbc", key, iv);
      dataStr = decipher.update(encryptedData, "base64", "utf-8");
      dataStr += decipher.final("utf-8");
    }
    if (signedState.compressed) {
      const decompressed = gunzipSync(Buffer.from(dataStr, "base64"));
      dataStr = decompressed.toString("utf-8");
    }
    return JSON.parse(dataStr);
  }
  getBackups(componentId) {
    return (this.stateBackups.get(componentId) || []).map((b) => b.signedState);
  }
  getLatestBackup(componentId) {
    const backups = this.stateBackups.get(componentId);
    if (!backups || backups.length === 0) return null;
    return backups[backups.length - 1].signedState;
  }
  backupState(componentId, signedState) {
    if (!this.stateBackups.has(componentId)) {
      this.stateBackups.set(componentId, []);
    }
    const backups = this.stateBackups.get(componentId);
    backups.push({ signedState, backedUpAt: Date.now() });
    while (backups.length > this.config.maxBackups) {
      backups.shift();
    }
  }
  computeSignature(signedState) {
    return this.computeSignatureWithKey(signedState, this.secret);
  }
  computeSignatureWithKey(signedState, key) {
    const payload = `${signedState.componentId}:${signedState.version}:${signedState.timestamp}:${signedState.data}${signedState.nonce ? ":" + signedState.nonce : ""}`;
    return createHmac("sha256", key).update(payload).digest("hex");
  }
  timingSafeEqual(a, b) {
    if (a.length !== b.length) return false;
    const bufA = Buffer.from(a);
    const bufB = Buffer.from(b);
    try {
      const { timingSafeEqual: tse } = __require("crypto");
      return tse(bufA, bufB);
    } catch {
      return a === b;
    }
  }
  deriveEncryptionKey() {
    if (this.cachedEncryptionKey) return this.cachedEncryptionKey;
    this.cachedEncryptionKey = scryptSync(this.secret, this.encryptionSalt, 32);
    return this.cachedEncryptionKey;
  }
  setupKeyRotation() {
    this.rotationTimer = setInterval(() => {
      this.previousSecrets.unshift(this.secret);
      if (this.previousSecrets.length > 3) {
        this.previousSecrets.pop();
      }
      this.secret = randomBytes(32);
      this.cachedEncryptionKey = null;
      liveLog("state", null, "Key rotation completed");
    }, this.config.rotationInterval);
  }
  /** Remove nonces older than nonceTTL + 10s from the replay detection map. */
  cleanupNonces() {
    const cutoff = Date.now() - (this.config.nonceTTL + 10 * 1e3);
    for (const [nonce, ts] of this.usedNonces) {
      if (ts < cutoff) this.usedNonces.delete(nonce);
    }
  }
  shutdown() {
    if (this.rotationTimer) clearInterval(this.rotationTimer);
    if (this.nonceCleanupTimer) clearInterval(this.nonceCleanupTimer);
  }
};
var PerformanceMonitor = class extends EventEmitter {
  components = /* @__PURE__ */ new Map();
  alerts = [];
  config;
  constructor(config = {}) {
    super();
    this.config = {
      slowActionThreshold: config.slowActionThreshold ?? 1e3,
      slowRenderThreshold: config.slowRenderThreshold ?? 500,
      highErrorRateThreshold: config.highErrorRateThreshold ?? 10,
      alertsEnabled: config.alertsEnabled ?? true
    };
  }
  initializeComponent(componentId, componentName) {
    this.components.set(componentId, {
      componentId,
      componentName,
      mountTime: Date.now(),
      actionTimes: /* @__PURE__ */ new Map(),
      renderTimes: [],
      stateChanges: 0,
      errorCount: 0,
      lastActivity: Date.now()
    });
  }
  recordRenderTime(componentId, time) {
    const metrics = this.components.get(componentId);
    if (!metrics) return;
    metrics.renderTimes.push(time);
    metrics.lastActivity = Date.now();
    if (this.config.alertsEnabled && time > this.config.slowRenderThreshold) {
      this.addAlert({
        type: "slow_render",
        componentId,
        componentName: metrics.componentName,
        message: `Slow render: ${time}ms (threshold: ${this.config.slowRenderThreshold}ms)`,
        value: time,
        threshold: this.config.slowRenderThreshold,
        timestamp: Date.now()
      });
    }
  }
  recordActionTime(componentId, action, time, error) {
    const metrics = this.components.get(componentId);
    if (!metrics) return;
    if (!metrics.actionTimes.has(action)) {
      metrics.actionTimes.set(action, []);
    }
    metrics.actionTimes.get(action).push(time);
    metrics.lastActivity = Date.now();
    if (error) {
      metrics.errorCount++;
      if (this.config.alertsEnabled && metrics.errorCount >= this.config.highErrorRateThreshold) {
        this.addAlert({
          type: "high_error_rate",
          componentId,
          componentName: metrics.componentName,
          message: `High error rate: ${metrics.errorCount} errors`,
          value: metrics.errorCount,
          threshold: this.config.highErrorRateThreshold,
          timestamp: Date.now()
        });
      }
    }
    if (this.config.alertsEnabled && time > this.config.slowActionThreshold) {
      this.addAlert({
        type: "slow_action",
        componentId,
        componentName: metrics.componentName,
        message: `Slow action '${action}': ${time}ms (threshold: ${this.config.slowActionThreshold}ms)`,
        value: time,
        threshold: this.config.slowActionThreshold,
        timestamp: Date.now()
      });
    }
  }
  recordStateChange(componentId) {
    const metrics = this.components.get(componentId);
    if (metrics) {
      metrics.stateChanges++;
      metrics.lastActivity = Date.now();
    }
  }
  removeComponent(componentId) {
    this.components.delete(componentId);
  }
  getComponentMetrics(componentId) {
    return this.components.get(componentId) ?? null;
  }
  getAllMetrics() {
    return Array.from(this.components.values());
  }
  getAlerts(limit = 50) {
    return this.alerts.slice(-limit);
  }
  clearAlerts() {
    this.alerts = [];
  }
  getStats() {
    return {
      totalComponents: this.components.size,
      totalAlerts: this.alerts.length,
      components: Array.from(this.components.entries()).map(([id, m]) => ({
        id,
        name: m.componentName,
        renderCount: m.renderTimes.length,
        avgRenderTime: m.renderTimes.length > 0 ? m.renderTimes.reduce((a, b) => a + b, 0) / m.renderTimes.length : 0,
        actionCount: Array.from(m.actionTimes.values()).reduce((sum, times) => sum + times.length, 0),
        stateChanges: m.stateChanges,
        errorCount: m.errorCount
      }))
    };
  }
  addAlert(alert) {
    this.alerts.push(alert);
    if (this.alerts.length > 1e3) {
      this.alerts = this.alerts.slice(-500);
    }
    liveWarn("performance", alert.componentId, alert.message);
    this.emit("alert", alert);
  }
};

// src/upload/FileUploadManager.ts
var MAGIC_BYTES = {
  "image/jpeg": [{ bytes: [255, 216, 255] }],
  "image/png": [{ bytes: [137, 80, 78, 71, 13, 10, 26, 10] }],
  "image/gif": [
    { bytes: [71, 73, 70, 56, 55, 97] },
    { bytes: [71, 73, 70, 56, 57, 97] }
  ],
  "image/webp": [{ bytes: [82, 73, 70, 70], offset: 0 }],
  "application/pdf": [{ bytes: [37, 80, 68, 70] }],
  "application/zip": [
    { bytes: [80, 75, 3, 4] },
    { bytes: [80, 75, 5, 6] }
  ],
  "application/gzip": [{ bytes: [31, 139] }]
};
var FileUploadManager = class {
  activeUploads = /* @__PURE__ */ new Map();
  maxUploadSize;
  chunkTimeout;
  userUploadBytes = /* @__PURE__ */ new Map();
  maxBytesPerUser;
  quotaResetInterval;
  allowedTypes;
  blockedExtensions;
  uploadsDir;
  customAssembleFile;
  cleanupTimer;
  quotaTimer;
  constructor(config = {}) {
    this.maxUploadSize = config.maxUploadSize ?? 50 * 1024 * 1024;
    this.chunkTimeout = config.chunkTimeout ?? 3e4;
    this.maxBytesPerUser = config.maxBytesPerUser ?? 500 * 1024 * 1024;
    this.quotaResetInterval = config.quotaResetInterval ?? 24 * 60 * 60 * 1e3;
    this.allowedTypes = config.allowedTypes ?? [
      "image/jpeg",
      "image/png",
      "image/gif",
      "image/webp",
      "image/svg+xml",
      "application/pdf",
      "text/plain",
      "text/csv",
      "text/markdown",
      "application/json",
      "application/zip",
      "application/gzip"
    ];
    this.blockedExtensions = new Set(config.blockedExtensions ?? [
      ".exe",
      ".bat",
      ".cmd",
      ".com",
      ".msi",
      ".scr",
      ".pif",
      ".sh",
      ".bash",
      ".zsh",
      ".csh",
      ".ps1",
      ".psm1",
      ".psd1",
      ".vbs",
      ".vbe",
      ".js",
      ".jse",
      ".wsf",
      ".wsh",
      ".dll",
      ".sys",
      ".drv",
      ".so",
      ".dylib"
    ]);
    this.uploadsDir = config.uploadsDir ?? "./uploads";
    this.customAssembleFile = config.assembleFile;
    this.cleanupTimer = setInterval(() => this.cleanupStaleUploads(), 5 * 60 * 1e3);
    this.quotaTimer = setInterval(() => this.resetUploadQuotas(), this.quotaResetInterval);
  }
  async startUpload(message, userId) {
    try {
      const { uploadId, componentId, filename, fileType, fileSize, chunkSize = 64 * 1024 } = message;
      if (fileSize > this.maxUploadSize) {
        throw new Error(`File too large: ${fileSize} bytes. Max: ${this.maxUploadSize} bytes`);
      }
      if (userId) {
        const currentUsage = this.userUploadBytes.get(userId) || 0;
        if (currentUsage + fileSize > this.maxBytesPerUser) {
          throw new Error(`Upload quota exceeded for user.`);
        }
      }
      if (this.allowedTypes.length > 0 && !this.allowedTypes.includes(fileType)) {
        throw new Error(`File type not allowed: ${fileType}`);
      }
      const pathModule = await import('path');
      const safeBase = pathModule.basename(filename);
      const ext = pathModule.extname(safeBase).toLowerCase();
      if (this.blockedExtensions.has(ext)) {
        throw new Error(`File extension not allowed: ${ext}`);
      }
      const parts = safeBase.split(".");
      if (parts.length > 2) {
        for (let i = 1; i < parts.length - 1; i++) {
          const intermediateExt = "." + parts[i].toLowerCase();
          if (this.blockedExtensions.has(intermediateExt)) {
            throw new Error(`Suspicious double extension detected: ${intermediateExt} in ${safeBase}`);
          }
        }
      }
      if (safeBase.length > 255) {
        throw new Error("Filename too long");
      }
      if (this.activeUploads.has(uploadId)) {
        throw new Error(`Upload ${uploadId} already in progress`);
      }
      const totalChunks = Math.ceil(fileSize / chunkSize);
      const upload = {
        uploadId,
        componentId,
        filename,
        fileType,
        fileSize,
        totalChunks,
        receivedChunks: /* @__PURE__ */ new Map(),
        bytesReceived: 0,
        startTime: Date.now(),
        lastChunkTime: Date.now()
      };
      this.activeUploads.set(uploadId, upload);
      if (userId) {
        const currentUsage = this.userUploadBytes.get(userId) || 0;
        this.userUploadBytes.set(userId, currentUsage + fileSize);
      }
      liveLog("messages", componentId, `Upload started: ${uploadId} (${filename}, ${fileSize} bytes)`);
      return { success: true };
    } catch (error) {
      return { success: false, error: error.message };
    }
  }
  async receiveChunk(message, binaryData = null) {
    const { uploadId, chunkIndex, totalChunks, data } = message;
    const upload = this.activeUploads.get(uploadId);
    if (!upload) throw new Error(`Upload ${uploadId} not found`);
    if (chunkIndex < 0 || chunkIndex >= totalChunks) {
      throw new Error(`Invalid chunk index: ${chunkIndex}`);
    }
    if (!upload.receivedChunks.has(chunkIndex)) {
      let chunkBytes;
      if (binaryData) {
        upload.receivedChunks.set(chunkIndex, binaryData);
        chunkBytes = binaryData.length;
      } else {
        upload.receivedChunks.set(chunkIndex, data);
        chunkBytes = Buffer.from(data, "base64").length;
      }
      upload.lastChunkTime = Date.now();
      upload.bytesReceived += chunkBytes;
    }
    const progress = upload.bytesReceived / upload.fileSize * 100;
    return {
      type: "FILE_UPLOAD_PROGRESS",
      componentId: upload.componentId,
      uploadId: upload.uploadId,
      chunkIndex,
      totalChunks,
      bytesUploaded: Math.min(upload.bytesReceived, upload.fileSize),
      totalBytes: upload.fileSize,
      progress: Math.min(progress, 100),
      timestamp: Date.now()
    };
  }
  async completeUpload(message) {
    try {
      const { uploadId } = message;
      const upload = this.activeUploads.get(uploadId);
      if (!upload) throw new Error(`Upload ${uploadId} not found`);
      if (upload.bytesReceived !== upload.fileSize) {
        throw new Error(`Incomplete upload: received ${upload.bytesReceived}/${upload.fileSize} bytes`);
      }
      this.validateContentMagicBytes(upload);
      const fileUrl = this.customAssembleFile ? await this.customAssembleFile(upload) : await this.defaultAssembleFile(upload);
      this.activeUploads.delete(uploadId);
      return {
        type: "FILE_UPLOAD_COMPLETE",
        componentId: upload.componentId,
        uploadId: upload.uploadId,
        success: true,
        filename: upload.filename,
        fileUrl,
        timestamp: Date.now()
      };
    } catch (error) {
      return {
        type: "FILE_UPLOAD_COMPLETE",
        componentId: "",
        uploadId: message.uploadId,
        success: false,
        error: error.message,
        timestamp: Date.now()
      };
    }
  }
  async defaultAssembleFile(upload) {
    const { writeFile, mkdir } = await import('fs/promises');
    const { existsSync } = await import('fs');
    const { join, extname, basename } = await import('path');
    if (!existsSync(this.uploadsDir)) {
      await mkdir(this.uploadsDir, { recursive: true });
    }
    const extension = extname(basename(upload.filename)).toLowerCase();
    const safeFilename = `${crypto.randomUUID()}${extension}`;
    const filePath = join(this.uploadsDir, safeFilename);
    const chunks = [];
    for (let i = 0; i < upload.totalChunks; i++) {
      const chunkData = upload.receivedChunks.get(i);
      if (chunkData) {
        if (Buffer.isBuffer(chunkData)) {
          chunks.push(chunkData);
        } else {
          chunks.push(Buffer.from(chunkData, "base64"));
        }
      }
    }
    await writeFile(filePath, Buffer.concat(chunks));
    return `/uploads/${safeFilename}`;
  }
  validateContentMagicBytes(upload) {
    const expectedSignatures = MAGIC_BYTES[upload.fileType];
    if (!expectedSignatures) return;
    const firstChunk = upload.receivedChunks.get(0);
    if (!firstChunk) throw new Error("Cannot validate file content: first chunk missing");
    const headerBuffer = Buffer.isBuffer(firstChunk) ? firstChunk : Buffer.from(firstChunk, "base64");
    let matched = false;
    for (const sig of expectedSignatures) {
      const offset = sig.offset ?? 0;
      if (headerBuffer.length < offset + sig.bytes.length) continue;
      let sigMatches = true;
      for (let i = 0; i < sig.bytes.length; i++) {
        if (headerBuffer[offset + i] !== sig.bytes[i]) {
          sigMatches = false;
          break;
        }
      }
      if (sigMatches) {
        matched = true;
        break;
      }
    }
    if (!matched) {
      throw new Error(
        `File content does not match claimed type '${upload.fileType}'. The file may be disguised as a different format.`
      );
    }
  }
  cleanupStaleUploads() {
    const now = Date.now();
    for (const [uploadId, upload] of this.activeUploads) {
      if (now - upload.lastChunkTime > this.chunkTimeout * 2) {
        this.activeUploads.delete(uploadId);
        liveLog("messages", null, `Cleaned up stale upload: ${uploadId}`);
      }
    }
  }
  resetUploadQuotas() {
    this.userUploadBytes.clear();
  }
  getUserUploadUsage(userId) {
    const used = this.userUploadBytes.get(userId) || 0;
    return { used, limit: this.maxBytesPerUser, remaining: Math.max(0, this.maxBytesPerUser - used) };
  }
  getUploadStatus(uploadId) {
    return this.activeUploads.get(uploadId) || null;
  }
  getStats() {
    return {
      activeUploads: this.activeUploads.size,
      maxUploadSize: this.maxUploadSize,
      allowedTypes: this.allowedTypes
    };
  }
  shutdown() {
    if (this.cleanupTimer) clearInterval(this.cleanupTimer);
    if (this.quotaTimer) clearInterval(this.quotaTimer);
  }
};
var WebSocketConnectionManager = class extends EventEmitter {
  connections = /* @__PURE__ */ new Map();
  connectionMetrics = /* @__PURE__ */ new Map();
  connectionPools = /* @__PURE__ */ new Map();
  /** Reverse index: connectionId -> Set of poolIds for O(1) cleanup */
  connectionPoolIndex = /* @__PURE__ */ new Map();
  messageQueues = /* @__PURE__ */ new Map();
  healthCheckTimer;
  heartbeatTimer;
  config;
  loadBalancerIndex = 0;
  constructor(config) {
    super();
    this.config = {
      maxConnections: 1e4,
      connectionTimeout: 3e4,
      heartbeatInterval: 3e4,
      reconnectAttempts: 5,
      reconnectDelay: 1e3,
      maxReconnectDelay: 3e4,
      jitterFactor: 0.1,
      loadBalancing: "round-robin",
      healthCheckInterval: 6e4,
      messageQueueSize: 1e3,
      offlineQueueEnabled: true,
      ...config
    };
    this.setupHealthMonitoring();
    this.setupHeartbeat();
  }
  registerConnection(ws, connectionId, poolId) {
    if (this.connections.size >= this.config.maxConnections) {
      throw new Error("Maximum connections exceeded");
    }
    const metrics = {
      id: connectionId,
      connectedAt: /* @__PURE__ */ new Date(),
      lastActivity: /* @__PURE__ */ new Date(),
      messagesSent: 0,
      messagesReceived: 0,
      bytesTransferred: 0,
      latency: 0,
      status: "connected",
      errorCount: 0,
      reconnectCount: 0
    };
    this.connections.set(connectionId, ws);
    this.connectionMetrics.set(connectionId, metrics);
    if (poolId) {
      this.addToPool(connectionId, poolId);
    }
    this.messageQueues.set(connectionId, []);
    liveLog("websocket", null, `Connection registered: ${connectionId}`);
    this.emit("connectionRegistered", { connectionId, poolId });
  }
  addToPool(connectionId, poolId) {
    if (!this.connectionPools.has(poolId)) {
      this.connectionPools.set(poolId, /* @__PURE__ */ new Set());
    }
    this.connectionPools.get(poolId).add(connectionId);
    if (!this.connectionPoolIndex.has(connectionId)) {
      this.connectionPoolIndex.set(connectionId, /* @__PURE__ */ new Set());
    }
    this.connectionPoolIndex.get(connectionId).add(poolId);
  }
  removeFromPool(connectionId, poolId) {
    const pool = this.connectionPools.get(poolId);
    if (pool) {
      pool.delete(connectionId);
      if (pool.size === 0) this.connectionPools.delete(poolId);
    }
    this.connectionPoolIndex.get(connectionId)?.delete(poolId);
  }
  cleanupConnection(connectionId) {
    this.connections.delete(connectionId);
    this.connectionMetrics.delete(connectionId);
    this.messageQueues.delete(connectionId);
    const poolIds = this.connectionPoolIndex.get(connectionId);
    if (poolIds) {
      for (const poolId of poolIds) {
        const pool = this.connectionPools.get(poolId);
        if (pool) {
          pool.delete(connectionId);
          if (pool.size === 0) this.connectionPools.delete(poolId);
        }
      }
      this.connectionPoolIndex.delete(connectionId);
    }
  }
  getConnectionMetrics(connectionId) {
    return this.connectionMetrics.get(connectionId) || null;
  }
  getAllConnectionMetrics() {
    return Array.from(this.connectionMetrics.values());
  }
  getSystemStats() {
    const totalConnections = this.connections.size;
    const activeConnections = Array.from(this.connections.values()).filter((ws) => ws.readyState === 1).length;
    const totalPools = this.connectionPools.size;
    const totalQueuedMessages = Array.from(this.messageQueues.values()).reduce((sum, queue) => sum + queue.length, 0);
    return {
      totalConnections,
      activeConnections,
      totalPools,
      totalQueuedMessages,
      maxConnections: this.config.maxConnections,
      connectionUtilization: totalConnections / this.config.maxConnections * 100
    };
  }
  setupHealthMonitoring() {
    this.healthCheckTimer = setInterval(() => this.performHealthChecks(), this.config.healthCheckInterval);
  }
  setupHeartbeat() {
    this.heartbeatTimer = setInterval(() => {
      for (const [connectionId, ws] of this.connections) {
        if (ws.readyState === 1) {
          try {
            const wsAny = ws;
            if (typeof wsAny.ping === "function") {
              wsAny._pingTime = Date.now();
              wsAny.ping();
            }
          } catch {
          }
        }
      }
    }, this.config.heartbeatInterval);
  }
  async performHealthChecks() {
    const now = Date.now();
    const unhealthy = [];
    for (const [connectionId, metrics] of this.connectionMetrics) {
      const ws = this.connections.get(connectionId);
      if (!ws || ws.readyState !== 1) {
        unhealthy.push(connectionId);
        continue;
      }
      const timeSinceActivity = now - metrics.lastActivity.getTime();
      if (timeSinceActivity > this.config.heartbeatInterval * 2) {
        metrics.status = "disconnected";
      }
      if (metrics.errorCount > 10) {
        unhealthy.push(connectionId);
      }
    }
    for (const connectionId of unhealthy) {
      const ws = this.connections.get(connectionId);
      if (ws) {
        try {
          ws.close();
        } catch {
        }
      }
      this.cleanupConnection(connectionId);
    }
  }
  shutdown() {
    if (this.healthCheckTimer) clearInterval(this.healthCheckTimer);
    if (this.heartbeatTimer) clearInterval(this.heartbeatTimer);
    for (const [, ws] of this.connections) {
      try {
        ws.close();
      } catch {
      }
    }
    this.connections.clear();
    this.connectionMetrics.clear();
    this.connectionPools.clear();
    this.connectionPoolIndex.clear();
    this.messageQueues.clear();
  }
};

// src/component/context.ts
var _ctx = null;
function setLiveComponentContext(ctx) {
  _ctx = ctx;
}
function getLiveComponentContext() {
  if (!_ctx) throw new Error("@fluxstack/live: LiveServer.start() must be called before using LiveComponents");
  return _ctx;
}

// src/component/managers/ComponentStateManager.ts
var _forbiddenSetCache = /* @__PURE__ */ new WeakMap();
var ComponentStateManager = class {
  _state;
  _proxyState;
  _inStateChange = false;
  _idBytes = null;
  _deepDiff;
  _deepDiffDepth;
  componentId;
  ws;
  emitFn;
  onStateChangeFn;
  constructor(opts) {
    this.componentId = opts.componentId;
    this.ws = opts.ws;
    this.emitFn = opts.emitFn;
    this.onStateChangeFn = opts.onStateChangeFn;
    this._deepDiff = opts.deepDiff ?? false;
    this._deepDiffDepth = opts.deepDiffDepth ?? 3;
    this._state = this._deepDiff ? structuredClone(opts.initialState) : opts.initialState;
    this._proxyState = this.createStateProxy(this._state);
  }
  get rawState() {
    return this._state;
  }
  get proxyState() {
    return this._proxyState;
  }
  /** Guard flag — prevents infinite recursion in onStateChange */
  get inStateChange() {
    return this._inStateChange;
  }
  createStateProxy(state) {
    const self = this;
    return new Proxy(state, {
      set(target, prop, value) {
        const oldValue = target[prop];
        if (oldValue !== value) {
          target[prop] = value;
          const changes = { [prop]: value };
          self.emitFn("STATE_DELTA", { delta: changes });
          if (!self._inStateChange) {
            self._inStateChange = true;
            try {
              self.onStateChangeFn(changes);
            } catch (err) {
              console.error(`[${self.componentId}] onStateChange error:`, err?.message || err);
            } finally {
              self._inStateChange = false;
            }
          }
        }
        return true;
      },
      get(target, prop) {
        return target[prop];
      }
    });
  }
  setState(updates) {
    const newUpdates = typeof updates === "function" ? updates(this._state) : updates;
    let actualChanges;
    let hasChanges;
    if (this._deepDiff) {
      const diff = computeDeepDiff(
        this._state,
        newUpdates,
        0,
        this._deepDiffDepth
      );
      if (diff === null) return;
      actualChanges = diff;
      hasChanges = true;
    } else {
      actualChanges = {};
      hasChanges = false;
      for (const key of Object.keys(newUpdates)) {
        if (this._state[key] !== newUpdates[key]) {
          actualChanges[key] = newUpdates[key];
          hasChanges = true;
        }
      }
    }
    if (!hasChanges) return;
    if (this._deepDiff) {
      deepAssign(this._state, actualChanges);
    } else {
      Object.assign(this._state, actualChanges);
    }
    this.emitFn("STATE_DELTA", { delta: actualChanges });
    if (!this._inStateChange) {
      this._inStateChange = true;
      try {
        this.onStateChangeFn(actualChanges);
      } catch (err) {
        console.error(`[${this.componentId}] onStateChange error:`, err?.message || err);
      } finally {
        this._inStateChange = false;
      }
    }
  }
  sendBinaryDelta(delta, encoder2) {
    const actualChanges = {};
    let hasChanges = false;
    for (const key of Object.keys(delta)) {
      if (this._state[key] !== delta[key]) {
        actualChanges[key] = delta[key];
        hasChanges = true;
      }
    }
    if (!hasChanges) return;
    Object.assign(this._state, actualChanges);
    const payload = encoder2(actualChanges);
    if (!this._idBytes) {
      this._idBytes = new TextEncoder().encode(this.componentId);
    }
    const idBytes = this._idBytes;
    const frame = new Uint8Array(1 + 1 + idBytes.length + payload.length);
    frame[0] = 1;
    frame[1] = idBytes.length;
    frame.set(idBytes, 2);
    frame.set(payload, 2 + idBytes.length);
    if (this.ws && this.ws.readyState === 1) {
      this.ws.send(frame);
    }
  }
  setValue(payload) {
    const { key, value } = payload;
    const update = { [key]: value };
    this.setState(update);
    return { success: true, key, value };
  }
  getSerializableState() {
    return this._proxyState;
  }
  /**
   * Create getters/setters for each state property directly on `target`.
   * This allows `this.count` instead of `this.state.count` in subclasses.
   */
  applyDirectAccessors(target, constructorFn) {
    let forbidden = _forbiddenSetCache.get(constructorFn);
    if (!forbidden) {
      forbidden = /* @__PURE__ */ new Set([
        ...Object.keys(target),
        ...Object.getOwnPropertyNames(Object.getPrototypeOf(target)),
        "state",
        "_state",
        "ws",
        "id",
        "room",
        "userId",
        "broadcastToRoom",
        "$private",
        "_privateState",
        "$room",
        "$rooms",
        "roomType",
        "roomHandles",
        "joinedRooms",
        "roomEventUnsubscribers",
        // Internal manager fields
        "_stateManager",
        "_roomProxyManager",
        "_actionSecurity",
        "_messaging"
      ]);
      _forbiddenSetCache.set(constructorFn, forbidden);
    }
    for (const key of Object.keys(this._state)) {
      if (!forbidden.has(key)) {
        Object.defineProperty(target, key, {
          get: () => this._state[key],
          set: (value) => {
            this._proxyState[key] = value;
          },
          enumerable: true,
          configurable: true
        });
      }
    }
  }
  /** Release cached resources */
  cleanup() {
    this._idBytes = null;
  }
};

// src/component/managers/ComponentMessaging.ts
var EMIT_OVERRIDE_KEY = /* @__PURE__ */ Symbol.for("fluxstack:emitOverride");
var ComponentMessaging = class {
  constructor(ctx) {
    this.ctx = ctx;
  }
  emit(type, payload) {
    const override = this.ctx.getEmitOverride();
    if (override) {
      override(type, payload);
      return;
    }
    const message = {
      type,
      componentId: this.ctx.componentId,
      payload,
      timestamp: Date.now(),
      userId: this.ctx.getUserId(),
      room: this.ctx.getRoom()
    };
    if (this.ctx.ws) {
      queueWsMessage(this.ctx.ws, message);
    }
  }
  broadcast(type, payload, excludeCurrentUser = false) {
    const room = this.ctx.getRoom();
    if (!room) {
      liveWarn("rooms", this.ctx.componentId, `[${this.ctx.componentId}] Cannot broadcast '${type}' - no room set`);
      return;
    }
    const message = {
      type,
      payload,
      room,
      excludeUser: excludeCurrentUser ? this.ctx.getUserId() : void 0
    };
    liveLog("rooms", this.ctx.componentId, `[${this.ctx.componentId}] Broadcasting '${type}' to room '${room}'`);
    this.ctx.getBroadcastToRoom()(message);
  }
};

// src/component/managers/ActionSecurityManager.ts
var BLOCKED_ACTIONS = /* @__PURE__ */ new Set([
  "constructor",
  "destroy",
  "executeAction",
  "getSerializableState",
  "onMount",
  "onDestroy",
  "onConnect",
  "onDisconnect",
  "onStateChange",
  "onRoomJoin",
  "onRoomLeave",
  "onRehydrate",
  "onAction",
  "onClientJoin",
  "onClientLeave",
  "setState",
  "sendBinaryDelta",
  "emit",
  "broadcast",
  "broadcastToRoom",
  "createStateProxy",
  "createDirectStateAccessors",
  "generateId",
  "setAuthContext",
  "_resetAuthContext",
  "$auth",
  "$private",
  "_privateState",
  "$persistent",
  "_inStateChange",
  "$room",
  "$rooms",
  "subscribeToRoom",
  "unsubscribeFromRoom",
  "emitRoomEvent",
  "onRoomEvent",
  "emitRoomEventWithState"
]);
var ActionSecurityManager = class {
  _actionCalls = /* @__PURE__ */ new Map();
  async validateAndExecute(action, payload, ctx) {
    const { component, componentClass, componentId } = ctx;
    try {
      if (BLOCKED_ACTIONS.has(action)) {
        throw new Error(`Action '${action}' is not callable`);
      }
      if (action.startsWith("_") || action.startsWith("#")) {
        throw new Error(`Action '${action}' is not callable`);
      }
      const publicActions = componentClass.publicActions;
      if (!publicActions) {
        console.warn(`[SECURITY] Component '${componentClass.componentName || componentClass.name}' has no publicActions defined. All remote actions are blocked.`);
        throw new Error(`Action '${action}' is not callable - component has no publicActions defined`);
      }
      if (!publicActions.includes(action)) {
        const methodExists = typeof component[action] === "function";
        if (methodExists) {
          const name = componentClass.componentName || componentClass.name;
          throw new Error(
            `Action '${action}' exists on '${name}' but is not listed in publicActions. Add it to: static publicActions = [..., '${action}']`
          );
        }
        throw new Error(`Action '${action}' is not callable`);
      }
      const method = component[action];
      if (typeof method !== "function") {
        throw new Error(`Action '${action}' not found on component`);
      }
      if (Object.prototype.hasOwnProperty.call(Object.prototype, action)) {
        throw new Error(`Action '${action}' is not callable`);
      }
      const rateLimit = componentClass.actionRateLimit;
      if (rateLimit) {
        const now = Date.now();
        const key = rateLimit.perAction ? action : "*";
        let entry = this._actionCalls.get(key);
        if (!entry || now - entry.windowStart >= rateLimit.windowMs) {
          entry = { count: 0, windowStart: now };
          this._actionCalls.set(key, entry);
        }
        entry.count++;
        if (entry.count > rateLimit.maxCalls) {
          throw new Error(`Action rate limit exceeded (max ${rateLimit.maxCalls} calls per ${rateLimit.windowMs}ms)`);
        }
      }
      const schemas = componentClass.actionSchemas;
      if (schemas && schemas[action]) {
        const result2 = schemas[action].safeParse(payload);
        if (!result2.success) {
          const errorMsg = result2.error?.message || result2.error?.issues?.map((i) => i.message).join(", ") || "Invalid payload";
          throw new Error(`Action '${action}' payload validation failed: ${errorMsg}`);
        }
        payload = result2.data ?? payload;
      }
      let hookResult;
      try {
        hookResult = await component.onAction(action, payload);
      } catch (hookError) {
        ctx.emitFn("ERROR", {
          action,
          error: `Action '${action}' failed pre-validation`
        });
        throw hookError;
      }
      if (hookResult === false) {
        throw new Error(`Action '${action}' was cancelled`);
      }
      const result = await method.call(component, payload);
      return result;
    } catch (error) {
      if (!error.message?.includes("was cancelled") && !error.message?.includes("pre-validation")) {
        ctx.emitFn("ERROR", {
          action,
          error: error.message
        });
      }
      throw error;
    }
  }
};

// src/component/managers/ComponentRoomProxy.ts
var RESERVED_KEYS = /* @__PURE__ */ new Set([
  "id",
  "state",
  "join",
  "leave",
  "emit",
  "on",
  "setState",
  // Function internals (proxy wraps a function)
  "call",
  "apply",
  "bind",
  "prototype",
  "length",
  "name",
  "arguments",
  "caller",
  // Symbol keys
  Symbol.toPrimitive,
  Symbol.toStringTag,
  Symbol.hasInstance
]);
var TYPED_RESERVED_KEYS = /* @__PURE__ */ new Set([
  "id",
  "state",
  "meta",
  "join",
  "leave",
  "emit",
  "on",
  "setState",
  "memberCount",
  // Proxy internals
  "then",
  "toJSON",
  "valueOf",
  "toString",
  Symbol.toPrimitive,
  Symbol.toStringTag,
  Symbol.hasInstance
]);
function wrapWithStateProxy(target, getState, setState) {
  return new Proxy(target, {
    get(obj, prop, receiver) {
      if (RESERVED_KEYS.has(prop) || typeof prop === "symbol") {
        return Reflect.get(obj, prop, receiver);
      }
      const desc = Object.getOwnPropertyDescriptor(obj, prop);
      if (desc) return Reflect.get(obj, prop, receiver);
      if (prop in obj) return Reflect.get(obj, prop, receiver);
      const st = getState();
      return st?.[prop];
    },
    set(_obj, prop, value) {
      if (typeof prop === "symbol") return false;
      setState({ [prop]: value });
      return true;
    }
  });
}
var ComponentRoomProxy = class {
  roomEventUnsubscribers = [];
  joinedRooms = /* @__PURE__ */ new Set();
  roomHandles = /* @__PURE__ */ new Map();
  _roomProxy = null;
  _roomsCache = null;
  _cachedCtx = null;
  roomType = "default";
  room;
  componentId;
  ws;
  getCtx;
  setStateFn;
  _deepDiff;
  _deepDiffDepth;
  _serverOnlyState;
  constructor(rctx) {
    this.componentId = rctx.componentId;
    this.ws = rctx.ws;
    this.room = rctx.defaultRoom;
    this.getCtx = rctx.getCtx;
    this.setStateFn = rctx.setStateFn;
    this._deepDiff = rctx.deepDiff ?? true;
    this._deepDiffDepth = rctx.deepDiffDepth;
    this._serverOnlyState = rctx.serverOnlyState ?? false;
    if (this.room) {
      this.joinedRooms.add(this.room);
      this.ctx.roomManager.joinRoom(this.componentId, this.room, this.ws, void 0, { deepDiff: this._deepDiff, deepDiffDepth: this._deepDiffDepth, serverOnlyState: this._serverOnlyState });
    }
  }
  /** Lazy context resolution — cached after first access */
  get ctx() {
    if (!this._cachedCtx) {
      this._cachedCtx = this.getCtx();
    }
    return this._cachedCtx;
  }
  get $room() {
    if (this._roomProxy) return this._roomProxy;
    const self = this;
    const createHandle = (roomId) => {
      if (this.roomHandles.has(roomId)) {
        return this.roomHandles.get(roomId);
      }
      const handle = {
        get id() {
          return roomId;
        },
        get state() {
          return self.ctx.roomManager.getRoomState(roomId);
        },
        join: (initialState) => {
          if (self.joinedRooms.has(roomId)) return;
          self.joinedRooms.add(roomId);
          self._roomsCache = null;
          self.ctx.roomManager.joinRoom(self.componentId, roomId, self.ws, initialState, { deepDiff: self._deepDiff, deepDiffDepth: self._deepDiffDepth, serverOnlyState: self._serverOnlyState });
        },
        leave: () => {
          if (!self.joinedRooms.has(roomId)) return;
          self.joinedRooms.delete(roomId);
          self._roomsCache = null;
          self.ctx.roomManager.leaveRoom(self.componentId, roomId);
        },
        emit: (event, data) => {
          return self.ctx.roomManager.emitToRoom(roomId, event, data, self.componentId);
        },
        on: (event, handler) => {
          const unsubscribe = self.ctx.roomEvents.on(
            "room",
            roomId,
            event,
            self.componentId,
            handler
          );
          self.roomEventUnsubscribers.push(unsubscribe);
          return unsubscribe;
        },
        setState: (updates) => {
          self.ctx.roomManager.setRoomState(roomId, updates, self.componentId);
        }
      };
      const proxied = wrapWithStateProxy(
        handle,
        () => self.ctx.roomManager.getRoomState(roomId),
        (updates) => self.ctx.roomManager.setRoomState(roomId, updates, self.componentId)
      );
      this.roomHandles.set(roomId, proxied);
      return proxied;
    };
    const proxyFn = ((roomIdOrClass, instanceId) => {
      if (typeof roomIdOrClass === "function" && instanceId !== void 0) {
        return self.$typedRoom(roomIdOrClass, instanceId);
      }
      return createHandle(roomIdOrClass);
    });
    const defaultHandle = this.room ? createHandle(this.room) : null;
    Object.defineProperties(proxyFn, {
      id: { get: () => self.room },
      state: { get: () => defaultHandle?.state ?? {} },
      join: {
        value: (initialState) => {
          if (!defaultHandle) throw new Error("No default room set");
          defaultHandle.join(initialState);
        }
      },
      leave: {
        value: () => {
          if (!defaultHandle) throw new Error("No default room set");
          defaultHandle.leave();
        }
      },
      emit: {
        value: (event, data) => {
          if (!defaultHandle) throw new Error("No default room set");
          return defaultHandle.emit(event, data);
        }
      },
      on: {
        value: (event, handler) => {
          if (!defaultHandle) throw new Error("No default room set");
          return defaultHandle.on(event, handler);
        }
      },
      setState: {
        value: (updates) => {
          if (!defaultHandle) throw new Error("No default room set");
          defaultHandle.setState(updates);
        }
      }
    });
    const defaultRoom = this.room;
    const wrapped = defaultRoom ? wrapWithStateProxy(
      proxyFn,
      () => self.ctx.roomManager.getRoomState(defaultRoom),
      (updates) => self.ctx.roomManager.setRoomState(defaultRoom, updates, self.componentId)
    ) : proxyFn;
    this._roomProxy = wrapped;
    return wrapped;
  }
  /**
   * Get a typed room handle backed by a LiveRoom class.
   *
   * Usage: `this.$room(ChatRoom, 'lobby')` → typed handle with custom methods
   *
   * The returned handle exposes:
   * - `.id`, `.state`, `.meta`, `.memberCount` — framework properties
   * - `.join(payload?)`, `.leave()`, `.emit()`, `.on()`, `.setState()` — framework API
   * - Any custom method defined on the LiveRoom subclass (e.g. `.addMessage()`, `.ban()`)
   *
   * The compound room ID is `${roomClass.roomName}:${instanceId}`.
   */
  $typedRoom(roomClass, instanceId) {
    const roomId = `${roomClass.roomName}:${instanceId}`;
    const self = this;
    const cached = this.roomHandles.get(roomId);
    if (cached) return cached;
    const handle = {
      get id() {
        return roomId;
      },
      get state() {
        const instance = self.ctx.roomManager.getRoomInstance?.(roomId);
        return instance ? instance.state : self.ctx.roomManager.getRoomState(roomId);
      },
      get meta() {
        const instance = self.ctx.roomManager.getRoomInstance?.(roomId);
        if (!instance) throw new Error(`Room '${roomId}' not found or not backed by a LiveRoom class`);
        return instance.meta;
      },
      get memberCount() {
        return self.ctx.roomManager.getMemberCount?.(roomId) ?? 0;
      },
      join: (payload) => {
        if (self.joinedRooms.has(roomId)) return {};
        const result = self.ctx.roomManager.joinRoom(
          self.componentId,
          roomId,
          self.ws,
          void 0,
          void 0,
          { userId: void 0, payload }
        );
        if ("rejected" in result && result.rejected) {
          return result;
        }
        self.joinedRooms.add(roomId);
        self._roomsCache = null;
        sendImmediate(self.ws, JSON.stringify({
          type: "ROOM_JOINED",
          componentId: self.componentId,
          roomId,
          event: "$room:joined",
          data: { state: result.state },
          timestamp: Date.now()
        }));
        return {};
      },
      leave: () => {
        if (!self.joinedRooms.has(roomId)) return;
        self.joinedRooms.delete(roomId);
        self._roomsCache = null;
        self.ctx.roomManager.leaveRoom(self.componentId, roomId, "leave");
      },
      emit: ((event, data) => {
        return self.ctx.roomManager.emitToRoom(roomId, event, data, self.componentId);
      }),
      on: (event, handler) => {
        const unsubscribe = self.ctx.roomEvents.on(
          "room",
          roomId,
          event,
          self.componentId,
          handler
        );
        self.roomEventUnsubscribers.push(unsubscribe);
        return unsubscribe;
      },
      setState: (updates) => {
        self.ctx.roomManager.setRoomState(roomId, updates, self.componentId);
      }
    };
    const proxied = new Proxy(handle, {
      get(obj, prop, receiver) {
        if (TYPED_RESERVED_KEYS.has(prop) || typeof prop === "symbol") {
          return Reflect.get(obj, prop, receiver);
        }
        if (prop in obj) return Reflect.get(obj, prop, receiver);
        const instance = self.ctx.roomManager.getRoomInstance?.(roomId);
        if (instance && prop in instance) {
          const val = instance[prop];
          return typeof val === "function" ? val.bind(instance) : val;
        }
        return void 0;
      }
    });
    this.roomHandles.set(roomId, proxied);
    return proxied;
  }
  get $rooms() {
    if (this._roomsCache) return this._roomsCache;
    this._roomsCache = Array.from(this.joinedRooms);
    return this._roomsCache;
  }
  getJoinedRooms() {
    return this.joinedRooms;
  }
  emitRoomEvent(event, data, notifySelf = false) {
    if (!this.room) {
      liveWarn("rooms", this.componentId, `[${this.componentId}] Cannot emit room event '${event}' - no room set`);
      return 0;
    }
    const excludeId = notifySelf ? void 0 : this.componentId;
    const notified = this.ctx.roomEvents.emit(this.roomType, this.room, event, data, excludeId);
    liveLog("rooms", this.componentId, `[${this.componentId}] Room event '${event}' -> ${notified} components`);
    return notified;
  }
  onRoomEvent(event, handler) {
    if (!this.room) {
      liveWarn("rooms", this.componentId, `[${this.componentId}] Cannot subscribe to room event '${event}' - no room set`);
      return;
    }
    const unsubscribe = this.ctx.roomEvents.on(
      this.roomType,
      this.room,
      event,
      this.componentId,
      handler
    );
    this.roomEventUnsubscribers.push(unsubscribe);
    liveLog("rooms", this.componentId, `[${this.componentId}] Subscribed to room event '${event}'`);
  }
  emitRoomEventWithState(event, data, stateUpdates) {
    this.setStateFn(stateUpdates);
    return this.emitRoomEvent(event, data, false);
  }
  subscribeToRoom(roomId) {
    this.room = roomId;
  }
  unsubscribeFromRoom() {
    this.room = void 0;
  }
  destroy() {
    for (const unsubscribe of this.roomEventUnsubscribers) {
      unsubscribe();
    }
    this.roomEventUnsubscribers = [];
    if (this.joinedRooms.size > 0 && this._cachedCtx) {
      for (const roomId of this.joinedRooms) {
        this._cachedCtx.roomManager.leaveRoom(this.componentId, roomId);
      }
    }
    this.joinedRooms.clear();
    this.roomHandles.clear();
    this._roomProxy = null;
    this._roomsCache = null;
  }
};

// src/component/LiveComponent.ts
var LiveComponent = class {
  /** Component name for registry lookup - must be defined in subclasses */
  static componentName;
  /** Default state - must be defined in subclasses */
  static defaultState;
  /**
   * Per-component logging control. Silent by default.
   *
   * @example
   * static logging = true                           // all categories
   * static logging = ['lifecycle', 'messages']      // specific categories
   */
  static logging;
  /**
   * Component-level auth configuration.
   */
  static auth;
  /**
   * Per-action auth configuration.
   */
  static actionAuth;
  /**
   * Zod schemas for action payload validation.
   * When defined, payloads are validated before the action method is called.
   *
   * @example
   * static actionSchemas = {
   *   sendMessage: z.object({ text: z.string().max(500) }),
   *   updatePosition: z.object({ x: z.number(), y: z.number() }),
   * }
   */
  static actionSchemas;
  /**
   * Rate limit for action execution.
   * Prevents clients from spamming expensive operations.
   *
   * @example
   * static actionRateLimit = { maxCalls: 10, windowMs: 1000, perAction: true }
   */
  static actionRateLimit;
  /**
   * Data that survives HMR reloads.
   */
  static persistent;
  /**
   * When true, only ONE server-side instance exists for this component.
   * All clients share the same state.
   */
  static singleton;
  /**
   * Component behavior options.
   *
   * @example
   * static $options = { deepDiff: true }
   */
  static $options;
  id;
  state;
  // Proxy wrapper (getter delegates to _stateManager)
  ws;
  room;
  userId;
  broadcastToRoom = () => {
  };
  // Server-only private state (NEVER sent to client)
  _privateState = {};
  // Auth context (injected by registry during mount, immutable after first set)
  _authContext = ANONYMOUS_CONTEXT;
  _authContextSet = false;
  // Room type for typed events (override in subclass)
  roomType = "default";
  // Singleton emit override
  [EMIT_OVERRIDE_KEY] = null;
  // ===== Internal Managers (composition) =====
  _stateManager;
  _messaging;
  _actionSecurity;
  _roomProxyManager;
  static publicActions;
  constructor(initialState, ws, options) {
    this.id = this.generateId();
    const ctor = this.constructor;
    this.ws = ws;
    this.room = options?.room;
    this.userId = options?.userId;
    this._messaging = new ComponentMessaging({
      componentId: this.id,
      ws: this.ws,
      getUserId: () => this.userId,
      getRoom: () => this.room,
      getBroadcastToRoom: () => this.broadcastToRoom,
      getEmitOverride: () => this[EMIT_OVERRIDE_KEY]
    });
    this._stateManager = new ComponentStateManager({
      componentId: this.id,
      initialState: { ...ctor.defaultState, ...initialState },
      ws: this.ws,
      emitFn: (type, payload) => this._messaging.emit(type, payload),
      onStateChangeFn: (changes) => this.onStateChange(changes),
      deepDiff: ctor.$options?.deepDiff ?? false,
      deepDiffDepth: ctor.$options?.deepDiffDepth
    });
    this.state = this._stateManager.proxyState;
    this._actionSecurity = new ActionSecurityManager();
    this._roomProxyManager = new ComponentRoomProxy({
      componentId: this.id,
      ws: this.ws,
      defaultRoom: this.room,
      getCtx: () => getLiveComponentContext(),
      setStateFn: (updates) => this.setState(updates),
      deepDiff: ctor.$options?.roomDeepDiff,
      deepDiffDepth: ctor.$options?.deepDiffDepth,
      serverOnlyState: ctor.$options?.serverOnlyRoomState
    });
    this._stateManager.applyDirectAccessors(this, this.constructor);
  }
  // ========================================
  // $private - Server-Only State
  // ========================================
  get $private() {
    return this._privateState;
  }
  // ========================================
  // $room - Unified Room System
  // ========================================
  /**
   * Unified room accessor.
   *
   * Usage:
   * - `this.$room` — default room handle (legacy)
   * - `this.$room('roomId')` — untyped room handle (legacy)
   * - `this.$room(ChatRoom, 'lobby')` — typed handle with custom methods
   */
  get $room() {
    return this._roomProxyManager.$room;
  }
  /**
   * List of room IDs this component is participating in.
   * Cached — invalidated on join/leave.
   */
  get $rooms() {
    return this._roomProxyManager.$rooms;
  }
  // ========================================
  // $auth - Authentication Context
  // ========================================
  get $auth() {
    return this._authContext;
  }
  /** @internal - Immutable after first set to prevent privilege escalation */
  setAuthContext(context) {
    if (this._authContextSet) {
      throw new Error("Auth context is immutable after initial set");
    }
    this._authContext = context;
    this._authContextSet = true;
    if (context.authenticated && context.user?.id && !this.userId) {
      this.userId = context.user.id;
    }
  }
  /** @internal - Reset auth context (for registry use in reconnection) */
  _resetAuthContext() {
    this._authContextSet = false;
    this._authContext = ANONYMOUS_CONTEXT;
  }
  // ========================================
  // $persistent - HMR-Safe State
  // ========================================
  get $persistent() {
    const ctor = this.constructor;
    const name = ctor.componentName || ctor.name;
    const key = `__fluxstack_persistent_${name}`;
    if (!globalThis[key]) {
      globalThis[key] = { ...ctor.persistent || {} };
    }
    return globalThis[key];
  }
  // ========================================
  // Lifecycle Hooks
  // ========================================
  onConnect() {
  }
  onMount() {
  }
  onDisconnect() {
  }
  onDestroy() {
  }
  onStateChange(changes) {
  }
  onRoomJoin(roomId) {
  }
  onRoomLeave(roomId) {
  }
  onRehydrate(previousState) {
  }
  onAction(action, payload) {
  }
  onClientJoin(connectionId, connectionCount) {
  }
  onClientLeave(connectionId, connectionCount) {
  }
  // ========================================
  // State Management (delegates to _stateManager)
  // ========================================
  setState(updates) {
    this._stateManager.setState(updates);
  }
  /**
   * Send a binary-encoded state delta directly over WebSocket.
   * Updates internal state (same as setState) then sends the encoder's output
   * as a binary frame: [0x01][idLen:u8][id_bytes:utf8][payload_bytes].
   * Bypasses the JSON batcher — ideal for high-frequency updates.
   */
  sendBinaryDelta(delta, encoder2) {
    this._stateManager.sendBinaryDelta(delta, encoder2);
  }
  setValue(payload) {
    return this._stateManager.setValue(payload);
  }
  // ========================================
  // Action Execution (delegates to _actionSecurity)
  // ========================================
  async executeAction(action, payload) {
    return this._actionSecurity.validateAndExecute(action, payload, {
      component: this,
      componentClass: this.constructor,
      componentId: this.id,
      emitFn: (type, p) => this.emit(type, p)
    });
  }
  // ========================================
  // Messaging (delegates to _messaging)
  // ========================================
  emit(type, payload) {
    this._messaging.emit(type, payload);
  }
  broadcast(type, payload, excludeCurrentUser = false) {
    this._messaging.broadcast(type, payload, excludeCurrentUser);
  }
  // ========================================
  // Room Events (delegates to _roomProxyManager)
  // ========================================
  emitRoomEvent(event, data, notifySelf = false) {
    return this._roomProxyManager.emitRoomEvent(event, data, notifySelf);
  }
  onRoomEvent(event, handler) {
    this._roomProxyManager.onRoomEvent(event, handler);
  }
  emitRoomEventWithState(event, data, stateUpdates) {
    return this._roomProxyManager.emitRoomEventWithState(event, data, stateUpdates);
  }
  subscribeToRoom(roomId) {
    this._roomProxyManager.subscribeToRoom(roomId);
    this.room = roomId;
  }
  unsubscribeFromRoom() {
    this._roomProxyManager.unsubscribeFromRoom();
    this.room = void 0;
  }
  // ========================================
  // Internal
  // ========================================
  generateId() {
    return `live-${crypto.randomUUID()}`;
  }
  destroy() {
    try {
      this.onDestroy();
    } catch (err) {
      console.error(`[${this.id}] onDestroy error:`, err?.message || err);
    }
    this._roomProxyManager.destroy();
    this._stateManager.cleanup();
    this._privateState = {};
    this.room = void 0;
  }
  getSerializableState() {
    return this._stateManager.getSerializableState();
  }
};

// src/component/ComponentRegistry.ts
var ComponentRegistry = class {
  components = /* @__PURE__ */ new Map();
  definitions = /* @__PURE__ */ new Map();
  metadata = /* @__PURE__ */ new Map();
  rooms = /* @__PURE__ */ new Map();
  wsConnections = /* @__PURE__ */ new Map();
  autoDiscoveredComponents = /* @__PURE__ */ new Map();
  healthCheckInterval;
  singletons = /* @__PURE__ */ new Map();
  remoteSingletons = /* @__PURE__ */ new Map();
  cluster;
  authManager;
  stateSignature;
  performanceMonitor;
  constructor(deps) {
    this.authManager = deps.authManager;
    this.stateSignature = deps.stateSignature;
    this.performanceMonitor = deps.performanceMonitor;
    this.cluster = deps.cluster;
    this.setupHealthMonitoring();
    this.setupClusterHandlers();
  }
  /** Set up handlers for incoming cluster messages (deltas, forwarded actions). */
  setupClusterHandlers() {
    if (!this.cluster) return;
    this.cluster.onDelta((componentId, componentName, delta, sourceInstanceId) => {
      const remote = this.remoteSingletons.get(componentName);
      if (!remote || remote.componentId !== componentId) return;
      if (delta && remote.lastState) {
        Object.assign(remote.lastState, delta);
      }
      const message = JSON.stringify({
        type: "STATE_DELTA",
        componentId,
        payload: { delta },
        timestamp: Date.now()
      });
      const dead = [];
      for (const [connId, ws] of remote.connections) {
        if (ws.readyState === 1) {
          try {
            ws.send(message);
          } catch {
            dead.push(connId);
          }
        } else {
          dead.push(connId);
        }
      }
      for (const connId of dead) remote.connections.delete(connId);
    });
    this.cluster.onOwnershipLost((componentName) => {
      const singleton = this.singletons.get(componentName);
      if (!singleton) return;
      this.cluster.saveSingletonState(componentName, singleton.instance.getSerializableState()).catch(() => {
      });
      const errorMsg = JSON.stringify({
        type: "ERROR",
        componentId: singleton.instance.id,
        payload: { error: "OWNERSHIP_LOST: singleton moved to another server" },
        timestamp: Date.now()
      });
      for (const [, ws] of singleton.connections) {
        try {
          ws.send(errorMsg);
        } catch {
        }
      }
      this.cleanupComponent(singleton.instance.id);
      this.singletons.delete(componentName);
    });
    this.cluster.onActionForward(async (request) => {
      try {
        const stillOwner = await this.cluster.verifySingletonOwnership(request.componentName);
        if (!stillOwner) {
          return { success: false, error: "OWNERSHIP_LOST: this instance no longer owns the singleton", requestId: request.requestId };
        }
        const result = await this.executeAction(request.componentId, request.action, request.payload);
        return { success: true, result, requestId: request.requestId };
      } catch (error) {
        return { success: false, error: error.message, requestId: request.requestId };
      }
    });
  }
  setupHealthMonitoring() {
    this.healthCheckInterval = setInterval(() => this.performHealthChecks(), 3e4);
  }
  registerComponent(definition) {
    this.definitions.set(definition.name, definition);
    liveLog("lifecycle", null, `Registered component: ${definition.name}`);
  }
  registerComponentClass(name, componentClass) {
    this.autoDiscoveredComponents.set(name, componentClass);
  }
  async autoDiscoverComponents(componentsPath) {
    try {
      const fs = await import('fs');
      const path = await import('path');
      if (!fs.existsSync(componentsPath)) return;
      const files = fs.readdirSync(componentsPath);
      for (const file of files) {
        if (file.endsWith(".ts") || file.endsWith(".js")) {
          try {
            const fullPath = path.join(componentsPath, file);
            const module = await import(fullPath);
            Object.keys(module).forEach((exportName) => {
              const exportedItem = module[exportName];
              if (typeof exportedItem === "function" && exportedItem.prototype && this.isLiveComponentClass(exportedItem)) {
                const componentName = exportedItem.componentName || exportName.replace(/Component$/, "");
                this.registerComponentClass(componentName, exportedItem);
                liveLog("lifecycle", null, `Auto-discovered component: ${componentName} (from ${file})`);
              }
            });
          } catch {
          }
        }
      }
    } catch (error) {
      console.error("Auto-discovery failed:", error);
    }
  }
  isLiveComponentClass(cls) {
    try {
      if (typeof cls.componentName === "string") return true;
      if (cls.prototype && typeof cls.prototype.executeAction === "function" && typeof cls.prototype.setState === "function" && typeof cls.prototype.getSerializableState === "function") return true;
      let prototype = cls.prototype;
      while (prototype) {
        const name = prototype.constructor.name;
        if (name === "LiveComponent" || name === "_LiveComponent") return true;
        prototype = Object.getPrototypeOf(prototype);
      }
      return false;
    } catch {
      return false;
    }
  }
  async mountComponent(ws, componentName, props = {}, options) {
    const startTime = Date.now();
    try {
      const definition = this.definitions.get(componentName);
      let ComponentClass = null;
      let initialState = {};
      if (definition) {
        ComponentClass = definition.component;
        initialState = definition.initialState;
      } else {
        ComponentClass = this.autoDiscoveredComponents.get(componentName) ?? null;
        if (!ComponentClass) {
          const variations = [
            componentName + "Component",
            componentName.charAt(0).toUpperCase() + componentName.slice(1) + "Component",
            componentName.charAt(0).toUpperCase() + componentName.slice(1)
          ];
          for (const variation of variations) {
            ComponentClass = this.autoDiscoveredComponents.get(variation) ?? null;
            if (ComponentClass) break;
          }
        }
        if (!ComponentClass) throw new Error(`Component '${componentName}' not found`);
        initialState = {};
      }
      const authContext = ws.data?.authContext || ANONYMOUS_CONTEXT;
      const componentAuth = ComponentClass.auth;
      const authResult = this.authManager.authorizeComponent(authContext, componentAuth);
      if (!authResult.allowed) throw new Error(`AUTH_DENIED: ${authResult.reason}`);
      const isSingleton = ComponentClass.singleton === true;
      let clusterSingletonId = null;
      if (isSingleton) {
        const existing = this.singletons.get(componentName);
        if (existing) {
          const connId = ws.data?.connectionId || `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
          existing.connections.set(connId, ws);
          this.ensureWsData(ws, options?.userId);
          ws.data.components.set(existing.instance.id, existing.instance);
          const signedState2 = this.stateSignature.signState(existing.instance.id, {
            ...existing.instance.getSerializableState(),
            __componentName: componentName
          }, 1, { compress: true, backup: true });
          sendImmediate(ws, JSON.stringify({
            type: "STATE_UPDATE",
            componentId: existing.instance.id,
            payload: { state: existing.instance.getSerializableState(), signedState: signedState2 },
            timestamp: Date.now()
          }));
          try {
            existing.instance.onClientJoin(connId, existing.connections.size);
          } catch {
          }
          return { componentId: existing.instance.id, initialState: existing.instance.getSerializableState(), signedState: signedState2 };
        }
        const existingRemote = this.remoteSingletons.get(componentName);
        if (existingRemote) {
          const connId = ws.data?.connectionId || `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
          this.ensureWsData(ws, options?.userId);
          existingRemote.connections.set(connId, ws);
          sendImmediate(ws, JSON.stringify({
            type: "STATE_UPDATE",
            componentId: existingRemote.componentId,
            payload: { state: existingRemote.lastState },
            timestamp: Date.now()
          }));
          return { componentId: existingRemote.componentId, initialState: existingRemote.lastState, signedState: null };
        }
        if (this.cluster) {
          clusterSingletonId = `live-${crypto.randomUUID()}`;
          const claim = await this.cluster.claimSingleton(componentName, clusterSingletonId);
          if (!claim.claimed) {
            clusterSingletonId = null;
            const owner = await this.cluster.getSingletonOwner(componentName);
            if (owner) {
              const stored = await this.cluster.loadState(owner.componentId);
              const connId = ws.data?.connectionId || `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
              this.ensureWsData(ws, options?.userId);
              const remote = {
                componentName,
                componentId: owner.componentId,
                ownerInstanceId: owner.instanceId,
                lastState: stored?.state || {},
                connections: /* @__PURE__ */ new Map([[connId, ws]])
              };
              this.remoteSingletons.set(componentName, remote);
              sendImmediate(ws, JSON.stringify({
                type: "STATE_UPDATE",
                componentId: owner.componentId,
                payload: { state: remote.lastState },
                timestamp: Date.now()
              }));
              return { componentId: owner.componentId, initialState: remote.lastState, signedState: null };
            }
          }
          if (claim.recoveredState) {
            props = { ...props, ...claim.recoveredState };
          }
        }
      }
      const component = new ComponentClass({ ...initialState, ...props }, ws, options);
      component.setAuthContext(authContext);
      if (clusterSingletonId) {
        ;
        component.id = clusterSingletonId;
      }
      component.broadcastToRoom = (message) => {
        this.broadcastToRoom(message, component.id);
      };
      const metadata = this.createComponentMetadata(component.id, componentName, options?.version);
      this.metadata.set(component.id, metadata);
      this.components.set(component.id, component);
      this.wsConnections.set(component.id, ws);
      if (options?.room) this.subscribeToRoom(component.id, options.room);
      this.ensureWsData(ws, options?.userId);
      ws.data.components.set(component.id, component);
      if (isSingleton) {
        const connId = ws.data.connectionId || `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
        const connections = /* @__PURE__ */ new Map();
        connections.set(connId, ws);
        this.singletons.set(componentName, { instance: component, connections });
        if (this.cluster) {
          this.cluster.saveState(component.id, componentName, component.getSerializableState()).catch(() => {
          });
          this.cluster.saveSingletonState(componentName, component.getSerializableState()).catch(() => {
          });
        }
        ;
        component[EMIT_OVERRIDE_KEY] = (type, payload) => {
          const message = {
            type,
            componentId: component.id,
            payload,
            timestamp: Date.now(),
            userId: component.userId,
            room: component.room
          };
          const serialized = JSON.stringify(message);
          const singleton = this.singletons.get(componentName);
          if (singleton) {
            const dead = [];
            for (const [cId, cWs] of singleton.connections) {
              try {
                cWs.send(serialized);
              } catch {
                dead.push(cId);
              }
            }
            for (const cId of dead) singleton.connections.delete(cId);
          }
          if (this.cluster && type === "STATE_DELTA" && payload?.delta) {
            this.cluster.publishDelta(component.id, componentName, payload.delta).catch(() => {
            });
            this.cluster.saveState(component.id, componentName, component.getSerializableState()).catch(() => {
            });
            this.cluster.saveSingletonState(componentName, component.getSerializableState()).catch(() => {
            });
          }
        };
        try {
          component.onClientJoin(connId, 1);
        } catch {
        }
      }
      metadata.state = "active";
      const renderTime = Date.now() - startTime;
      this.recordComponentMetrics(component.id, renderTime);
      registerComponentLogging(component.id, ComponentClass.logging);
      this.performanceMonitor.initializeComponent(component.id, componentName);
      this.performanceMonitor.recordRenderTime(component.id, renderTime);
      const signedState = this.stateSignature.signState(component.id, {
        ...component.getSerializableState(),
        __componentName: componentName
      }, 1, { compress: true, backup: true });
      component.emit("STATE_UPDATE", {
        state: component.getSerializableState(),
        signedState
      });
      try {
        component.onConnect();
      } catch {
      }
      try {
        await component.onMount();
      } catch (err) {
        ;
        component.emit("ERROR", { action: "onMount", error: `Mount initialization failed: ${err?.message || err}` });
      }
      return { componentId: component.id, initialState: component.getSerializableState(), signedState };
    } catch (error) {
      console.error(`Failed to mount component ${componentName}:`, error);
      throw error;
    }
  }
  async rehydrateComponent(componentId, componentName, signedState, ws, options) {
    try {
      const validation = this.stateSignature.validateState(signedState, { skipNonce: true });
      if (!validation.valid) return { success: false, error: validation.error || "Invalid state signature" };
      const definition = this.definitions.get(componentName);
      let ComponentClass = null;
      let initialState = {};
      if (definition) {
        ComponentClass = definition.component;
        initialState = definition.initialState;
      } else {
        ComponentClass = this.autoDiscoveredComponents.get(componentName) ?? null;
        if (!ComponentClass) {
          const variations = [componentName + "Component", componentName.charAt(0).toUpperCase() + componentName.slice(1) + "Component", componentName.charAt(0).toUpperCase() + componentName.slice(1)];
          for (const variation of variations) {
            ComponentClass = this.autoDiscoveredComponents.get(variation) ?? null;
            if (ComponentClass) break;
          }
        }
        if (!ComponentClass) return { success: false, error: `Component '${componentName}' not found` };
      }
      const authContext = ws.data?.authContext || ANONYMOUS_CONTEXT;
      const componentAuth = ComponentClass.auth;
      const authResult = this.authManager.authorizeComponent(authContext, componentAuth);
      if (!authResult.allowed) return { success: false, error: `AUTH_DENIED: ${authResult.reason}` };
      const clientState = this.stateSignature.extractData(signedState);
      if (!clientState.__componentName || clientState.__componentName !== componentName) {
        return { success: false, error: "Component class mismatch - state tampering detected" };
      }
      const { __componentName, ...cleanState } = clientState;
      const finalState = definition ? { ...initialState, ...cleanState } : cleanState;
      const component = new ComponentClass(finalState, ws, options);
      component.setAuthContext(authContext);
      this.components.set(component.id, component);
      this.wsConnections.set(component.id, ws);
      if (options?.room) this.subscribeToRoom(component.id, options.room);
      this.ensureWsData(ws, options?.userId);
      ws.data.components.set(component.id, component);
      registerComponentLogging(component.id, ComponentClass.logging);
      const newSignedState = this.stateSignature.signState(
        component.id,
        { ...component.getSerializableState(), __componentName: componentName },
        signedState.version + 1
      );
      component.emit("STATE_REHYDRATED", {
        state: component.getSerializableState(),
        signedState: newSignedState,
        oldComponentId: componentId,
        newComponentId: component.id
      });
      try {
        component.onConnect();
      } catch {
      }
      try {
        component.onRehydrate(clientState);
      } catch {
      }
      try {
        await component.onMount();
      } catch {
      }
      return { success: true, newComponentId: component.id };
    } catch (error) {
      return { success: false, error: error.message };
    }
  }
  ensureWsData(ws, userId) {
    if (!ws.data) {
      ws.data = {
        connectionId: `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
        components: /* @__PURE__ */ new Map(),
        subscriptions: /* @__PURE__ */ new Set(),
        connectedAt: /* @__PURE__ */ new Date(),
        userId
      };
    }
    if (!ws.data.components) ws.data.components = /* @__PURE__ */ new Map();
  }
  isSingletonComponent(componentId) {
    for (const [, s] of this.singletons) if (s.instance.id === componentId) return true;
    return false;
  }
  removeSingletonConnection(componentId, connId, context = "unmount") {
    for (const [name, singleton] of this.singletons) {
      if (singleton.instance.id !== componentId) continue;
      if (connId) singleton.connections.delete(connId);
      if (singleton.connections.size === 0) {
        const finalState = singleton.instance.getSerializableState();
        try {
          singleton.instance.onDisconnect();
        } catch {
        }
        this.cleanupComponent(componentId);
        this.singletons.delete(name);
        if (this.cluster) {
          this.cluster.saveSingletonState(name, finalState).then(() => this.cluster.releaseSingleton(name)).then(() => this.cluster.deleteState(componentId)).catch(() => {
          });
        }
      }
      return true;
    }
    for (const [name, remote] of this.remoteSingletons) {
      if (remote.componentId !== componentId) continue;
      if (connId) remote.connections.delete(connId);
      if (remote.connections.size === 0) {
        this.remoteSingletons.delete(name);
      }
      return true;
    }
    return false;
  }
  unmountComponent(componentId, ws) {
    const component = this.components.get(componentId);
    if (!component) {
      if (ws) {
        const connId = ws.data?.connectionId;
        this.removeSingletonConnection(componentId, connId, "unmount");
      }
      return;
    }
    if (ws) {
      const connId = ws.data?.connectionId;
      ws.data?.components?.delete(componentId);
      if (this.isSingletonComponent(componentId)) {
        const singleton = this.singletons.get(this.getSingletonName(componentId) || "");
        const remaining = singleton ? singleton.connections.size - 1 : 0;
        try {
          component.onClientLeave(connId || "unknown", Math.max(0, remaining));
        } catch {
        }
      }
      if (this.removeSingletonConnection(componentId, connId, "unmount")) return;
    } else {
      if (this.removeSingletonConnection(componentId, void 0, "unmount")) return;
    }
    component.destroy?.();
    this.unsubscribeFromAllRooms(componentId);
    this.components.delete(componentId);
    this.wsConnections.delete(componentId);
    unregisterComponentLogging(componentId);
  }
  getSingletonName(componentId) {
    for (const [name, s] of this.singletons) {
      if (s.instance.id === componentId) return name;
    }
    return null;
  }
  /** Find a remote singleton entry by componentId. */
  findRemoteSingleton(componentId) {
    for (const [, entry] of this.remoteSingletons) {
      if (entry.componentId === componentId) return entry;
    }
    return null;
  }
  async executeAction(componentId, action, payload) {
    const component = this.components.get(componentId);
    if (!component) throw new Error(`COMPONENT_REHYDRATION_REQUIRED:${componentId}`);
    const componentClass = component.constructor;
    const actionAuthMap = componentClass.actionAuth;
    const actionAuth = actionAuthMap?.[action];
    if (actionAuth) {
      const authContext = component.$auth || ANONYMOUS_CONTEXT;
      const componentName = componentClass.componentName || componentClass.name;
      const authResult = await this.authManager.authorizeAction(authContext, componentName, action, actionAuth);
      if (!authResult.allowed) throw new Error(`AUTH_DENIED: ${authResult.reason}`);
    }
    return await component.executeAction?.(action, payload);
  }
  updateProperty(componentId, property, value) {
    const component = this.components.get(componentId);
    if (!component) throw new Error(`Component '${componentId}' not found`);
    component.setState?.({ [property]: value });
  }
  subscribeToRoom(componentId, roomId) {
    if (!this.rooms.has(roomId)) this.rooms.set(roomId, /* @__PURE__ */ new Set());
    this.rooms.get(roomId).add(componentId);
  }
  unsubscribeFromRoom(componentId, roomId) {
    const room = this.rooms.get(roomId);
    if (room) {
      room.delete(componentId);
      if (room.size === 0) this.rooms.delete(roomId);
    }
  }
  unsubscribeFromAllRooms(componentId) {
    for (const [roomId, components] of Array.from(this.rooms.entries())) {
      if (components.has(componentId)) this.unsubscribeFromRoom(componentId, roomId);
    }
  }
  broadcastToRoom(message, senderComponentId) {
    if (!message.room) return;
    const roomComponents = this.rooms.get(message.room);
    if (!roomComponents) return;
    const broadcastMessage = {
      type: "BROADCAST",
      componentId: senderComponentId || "system",
      payload: { type: message.type, data: message.payload },
      timestamp: Date.now(),
      room: message.room
    };
    for (const componentId of Array.from(roomComponents)) {
      const component = this.components.get(componentId);
      if (message.excludeUser && component?.userId === message.excludeUser) continue;
      const ws = this.wsConnections.get(componentId);
      if (ws) queueWsMessage(ws, broadcastMessage);
    }
  }
  async handleMessage(ws, message) {
    try {
      if (message.componentId) this.updateComponentActivity(message.componentId);
      switch (message.type) {
        case "COMPONENT_MOUNT":
          const mountResult = await this.mountComponent(ws, message.payload.component, message.payload.props, {
            room: message.payload.room,
            userId: message.userId,
            debugLabel: message.payload.debugLabel
          });
          return { success: true, result: mountResult };
        case "COMPONENT_UNMOUNT":
          this.unmountComponent(message.componentId, ws);
          return { success: true };
        case "CALL_ACTION": {
          const remoteSingleton = this.findRemoteSingleton(message.componentId);
          if (remoteSingleton && this.cluster) {
            const requestId = `req-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
            const request = {
              sourceInstanceId: this.cluster.instanceId,
              targetInstanceId: remoteSingleton.ownerInstanceId,
              componentId: remoteSingleton.componentId,
              componentName: remoteSingleton.componentName,
              action: message.action,
              payload: message.payload,
              requestId
            };
            const response = await this.cluster.forwardAction(request);
            if (!response.success) throw new Error(response.error || "Remote action failed");
            if (message.expectResponse) return { success: true, result: response.result };
            return null;
          }
          this.recordComponentMetrics(message.componentId, void 0, message.action);
          const actionStart = Date.now();
          try {
            const actionResult = await this.executeAction(message.componentId, message.action, message.payload);
            this.performanceMonitor.recordActionTime(message.componentId, message.action, Date.now() - actionStart);
            if (message.expectResponse) return { success: true, result: actionResult };
            return null;
          } catch (error) {
            this.performanceMonitor.recordActionTime(message.componentId, message.action, Date.now() - actionStart, error);
            throw error;
          }
        }
        case "PROPERTY_UPDATE":
          this.updateProperty(message.componentId, message.property, message.payload.value);
          return { success: true };
        default:
          return { success: false, error: "Unknown message type" };
      }
    } catch (error) {
      if (message.componentId) this.recordComponentError(message.componentId, error);
      return { success: false, error: error.message };
    }
  }
  cleanupConnection(ws) {
    if (!ws.data?.components) return;
    const componentsToCleanup = Array.from(ws.data.components.keys());
    const connId = ws.data.connectionId;
    for (const componentId of componentsToCleanup) {
      const component = this.components.get(componentId);
      if (component && !this.isSingletonComponent(componentId)) {
        try {
          component.onDisconnect();
        } catch {
        }
      }
      if (!this.removeSingletonConnection(componentId, connId || void 0, "disconnect")) {
        this.cleanupComponent(componentId);
      }
    }
    if (connId) {
      for (const [name, remote] of this.remoteSingletons) {
        remote.connections.delete(connId);
        if (remote.connections.size === 0) {
          this.remoteSingletons.delete(name);
        }
      }
    }
    ws.data.components.clear();
  }
  getStats() {
    return {
      components: this.components.size,
      definitions: this.definitions.size,
      rooms: this.rooms.size,
      connections: this.wsConnections.size,
      singletons: Object.fromEntries(
        Array.from(this.singletons.entries()).map(([name, s]) => [name, { componentId: s.instance.id, connections: s.connections.size }])
      ),
      remoteSingletons: Object.fromEntries(
        Array.from(this.remoteSingletons.entries()).map(([name, r]) => [name, { componentId: r.componentId, ownerInstanceId: r.ownerInstanceId, connections: r.connections.size }])
      ),
      roomDetails: Object.fromEntries(
        Array.from(this.rooms.entries()).map(([roomId, components]) => [roomId, components.size])
      )
    };
  }
  getRegisteredComponentNames() {
    return [.../* @__PURE__ */ new Set([...this.definitions.keys(), ...this.autoDiscoveredComponents.keys()])];
  }
  getComponent(componentId) {
    return this.components.get(componentId);
  }
  getRoomComponents(roomId) {
    const componentIds = this.rooms.get(roomId) || /* @__PURE__ */ new Set();
    return Array.from(componentIds).map((id) => this.components.get(id)).filter(Boolean);
  }
  createComponentMetadata(componentId, componentName, version = "1.0.0") {
    return {
      id: componentId,
      name: componentName,
      version,
      mountedAt: /* @__PURE__ */ new Date(),
      lastActivity: /* @__PURE__ */ new Date(),
      state: "mounting",
      healthStatus: "healthy",
      dependencies: [],
      services: /* @__PURE__ */ new Map(),
      metrics: { renderCount: 0, actionCount: 0, errorCount: 0, averageRenderTime: 0, memoryUsage: 0 },
      migrationHistory: []
    };
  }
  updateComponentActivity(componentId) {
    const metadata = this.metadata.get(componentId);
    if (metadata) {
      metadata.lastActivity = /* @__PURE__ */ new Date();
      metadata.state = "active";
      return true;
    }
    return false;
  }
  recordComponentMetrics(componentId, renderTime, action) {
    const metadata = this.metadata.get(componentId);
    if (!metadata) return;
    if (renderTime) {
      metadata.metrics.renderCount++;
      metadata.metrics.averageRenderTime = (metadata.metrics.averageRenderTime * (metadata.metrics.renderCount - 1) + renderTime) / metadata.metrics.renderCount;
      metadata.metrics.lastRenderTime = renderTime;
    }
    if (action) metadata.metrics.actionCount++;
    this.updateComponentActivity(componentId);
  }
  recordComponentError(componentId, error) {
    const metadata = this.metadata.get(componentId);
    if (metadata) {
      metadata.metrics.errorCount++;
      metadata.healthStatus = metadata.metrics.errorCount > 5 ? "unhealthy" : "degraded";
    }
  }
  performHealthChecks() {
    for (const [componentId, metadata] of this.metadata) {
      if (!this.components.get(componentId)) continue;
      if (metadata.metrics.errorCount > 10) metadata.healthStatus = "unhealthy";
      else if (Date.now() - metadata.lastActivity.getTime() > 3e5) metadata.healthStatus = "degraded";
    }
  }
  cleanupComponent(componentId) {
    const component = this.components.get(componentId);
    if (component) try {
      component.destroy?.();
    } catch {
    }
    this.performanceMonitor.removeComponent(componentId);
    unregisterComponentLogging(componentId);
    this.components.delete(componentId);
    this.metadata.delete(componentId);
    this.wsConnections.delete(componentId);
    for (const [roomId, componentIds] of this.rooms) {
      componentIds.delete(componentId);
      if (componentIds.size === 0) this.rooms.delete(roomId);
    }
  }
  cleanup() {
    if (this.healthCheckInterval) clearInterval(this.healthCheckInterval);
    this.singletons.clear();
    this.remoteSingletons.clear();
    for (const [componentId] of this.components) this.cleanupComponent(componentId);
  }
};

// src/connection/RateLimiter.ts
var ConnectionRateLimiter = class {
  tokens;
  lastRefill;
  maxTokens;
  refillRate;
  // tokens per second
  constructor(maxTokens = DEFAULT_RATE_LIMIT_MAX_TOKENS, refillRate = DEFAULT_RATE_LIMIT_REFILL_RATE) {
    this.maxTokens = maxTokens;
    this.tokens = maxTokens;
    this.refillRate = refillRate;
    this.lastRefill = Date.now();
  }
  tryConsume(count = 1) {
    this.refill();
    if (this.tokens >= count) {
      this.tokens -= count;
      return true;
    }
    return false;
  }
  refill() {
    const now = Date.now();
    const elapsed = (now - this.lastRefill) / 1e3;
    this.tokens = Math.min(this.maxTokens, this.tokens + elapsed * this.refillRate);
    this.lastRefill = now;
  }
};
var RateLimiterRegistry = class {
  limiters = /* @__PURE__ */ new Map();
  maxTokens;
  refillRate;
  constructor(maxTokens = DEFAULT_RATE_LIMIT_MAX_TOKENS, refillRate = DEFAULT_RATE_LIMIT_REFILL_RATE) {
    this.maxTokens = maxTokens;
    this.refillRate = refillRate;
  }
  /**
   * Get or create a rate limiter for a connection.
   */
  get(connectionId) {
    let limiter = this.limiters.get(connectionId);
    if (!limiter) {
      limiter = new ConnectionRateLimiter(this.maxTokens, this.refillRate);
      this.limiters.set(connectionId, limiter);
    }
    return limiter;
  }
  /**
   * Remove a rate limiter for a disconnected connection.
   */
  remove(connectionId) {
    this.limiters.delete(connectionId);
  }
};

// src/protocol/binary.ts
function encodeBinaryChunk(header, data) {
  const headerJson = JSON.stringify(header);
  const headerBuffer = Buffer.from(headerJson, "utf-8");
  const result = Buffer.alloc(4 + headerBuffer.length + data.length);
  result.writeUInt32LE(headerBuffer.length, 0);
  headerBuffer.copy(result, 4);
  Buffer.from(data).copy(result, 4 + headerBuffer.length);
  return result;
}
function decodeBinaryChunk(raw) {
  const buffer = raw instanceof ArrayBuffer ? Buffer.from(raw) : Buffer.from(raw.buffer, raw.byteOffset, raw.byteLength);
  const headerLength = buffer.readUInt32LE(0);
  const headerJson = buffer.slice(4, 4 + headerLength).toString("utf-8");
  const header = JSON.parse(headerJson);
  const data = buffer.slice(4 + headerLength);
  return { header, data };
}

// src/security/sanitize.ts
var DANGEROUS_KEYS = /* @__PURE__ */ new Set(["__proto__", "constructor", "prototype"]);
var MAX_DEPTH = 10;
function sanitizePayload(value, depth = 0) {
  if (depth > MAX_DEPTH) return value;
  if (Array.isArray(value)) {
    return value.map((item) => sanitizePayload(item, depth + 1));
  }
  if (value !== null && typeof value === "object") {
    const clean = {};
    for (const key of Object.keys(value)) {
      if (DANGEROUS_KEYS.has(key)) continue;
      clean[key] = sanitizePayload(value[key], depth + 1);
    }
    return clean;
  }
  return value;
}

// src/rooms/LiveRoom.ts
var LiveRoom = class {
  /** Unique room type name. Used as prefix in compound room IDs (e.g. "chat:lobby"). */
  static roomName;
  /** Initial public state template. Cloned per room instance. */
  static defaultState = {};
  /** Initial private metadata template. Cloned per room instance. */
  static defaultMeta = {};
  /** Room-level options */
  static $options;
  /** The unique room instance identifier (e.g. "chat:lobby") */
  id;
  /** Public state — synced to all connected clients via setState(). */
  state;
  /** Private metadata — NEVER leaves the server. Mutate directly. */
  meta;
  /** @internal Reference to the room manager for broadcasting */
  _manager;
  constructor(id, manager) {
    const ctor = this.constructor;
    this.id = id;
    this._manager = manager;
    this.state = structuredClone(ctor.defaultState ?? {});
    this.meta = structuredClone(ctor.defaultMeta ?? {});
  }
  // ===== Framework Methods =====
  /**
   * Update public state and broadcast changes to all room members.
   * Uses deep diff by default — only changed fields are sent over the wire.
   */
  setState(updates) {
    this._manager.setRoomState(this.id, updates);
  }
  /**
   * Emit a typed event to all members in this room.
   * @returns Number of members notified
   */
  emit(event, data) {
    return this._manager.emitToRoom(this.id, event, data);
  }
  /** Get current member count */
  get memberCount() {
    return this._manager.getMemberCount?.(this.id) ?? 0;
  }
  // ===== Lifecycle Hooks (override in subclass) =====
  /**
   * Called when a component attempts to join this room.
   * Return false to reject the join.
   */
  onJoin(_ctx2) {
  }
  /**
   * Called after a component leaves this room.
   */
  onLeave(_ctx2) {
  }
  /**
   * Called when an event is emitted to this room.
   * Can intercept/validate events before broadcasting.
   */
  onEvent(_event, _data, _ctx2) {
  }
  /**
   * Called once when the room is first created (first member joins).
   */
  onCreate() {
  }
  /**
   * Called when the last member leaves and the room is about to be destroyed.
   * Return false to keep the room alive (e.g., persist state).
   */
  onDestroy() {
  }
};

// src/rooms/RoomRegistry.ts
var RoomRegistry = class {
  rooms = /* @__PURE__ */ new Map();
  /**
   * Register a LiveRoom subclass.
   * @throws If the class doesn't define a static roomName
   */
  register(roomClass) {
    const name = roomClass.roomName;
    if (!name) {
      throw new Error("LiveRoom subclass must define static roomName");
    }
    if (this.rooms.has(name)) {
      throw new Error(`LiveRoom '${name}' is already registered`);
    }
    this.rooms.set(name, roomClass);
  }
  /**
   * Get a registered room class by name.
   */
  get(name) {
    return this.rooms.get(name);
  }
  /**
   * Check if a room class is registered.
   */
  has(name) {
    return this.rooms.has(name);
  }
  /**
   * Resolve a compound room ID (e.g. "chat:lobby") to its registered class.
   * Returns undefined if the prefix doesn't match any registered room.
   */
  resolveFromId(roomId) {
    const colonIdx = roomId.indexOf(":");
    if (colonIdx === -1) return void 0;
    const prefix = roomId.substring(0, colonIdx);
    return this.rooms.get(prefix);
  }
  /**
   * Get all registered room names.
   */
  getRegisteredNames() {
    return Array.from(this.rooms.keys());
  }
  /**
   * Check if a value is a LiveRoom subclass.
   */
  static isLiveRoomClass(cls) {
    if (typeof cls !== "function" || !cls.prototype) return false;
    if (typeof cls.roomName !== "string") return false;
    let proto = Object.getPrototypeOf(cls.prototype);
    while (proto) {
      if (proto.constructor === LiveRoom) return true;
      proto = Object.getPrototypeOf(proto);
    }
    return false;
  }
};

// src/server/LiveServer.ts
var LiveServer = class {
  // Public singletons (accessible for transport adapters & advanced usage)
  roomEvents;
  roomManager;
  authManager;
  stateSignature;
  performanceMonitor;
  fileUploadManager;
  connectionManager;
  registry;
  rateLimiter;
  roomRegistry;
  transport;
  options;
  constructor(options) {
    this.options = options;
    this.transport = options.transport;
    this.roomEvents = new RoomEventBus();
    this.roomManager = new LiveRoomManager(this.roomEvents, options.roomPubSub);
    this.authManager = new LiveAuthManager();
    this.stateSignature = new StateSignatureManager(options.stateSignature);
    this.performanceMonitor = new PerformanceMonitor(options.performance);
    this.fileUploadManager = new FileUploadManager(options.fileUpload);
    this.connectionManager = new WebSocketConnectionManager(options.connection);
    this.rateLimiter = new RateLimiterRegistry(options.rateLimitMaxTokens, options.rateLimitRefillRate);
    this.roomRegistry = new RoomRegistry();
    this.roomManager.roomRegistry = this.roomRegistry;
    if (options.rooms) {
      for (const roomClass of options.rooms) {
        this.roomRegistry.register(roomClass);
      }
    }
    this.registry = new ComponentRegistry({
      authManager: this.authManager,
      stateSignature: this.stateSignature,
      performanceMonitor: this.performanceMonitor,
      cluster: options.cluster
    });
    setLiveComponentContext({
      roomEvents: this.roomEvents,
      roomManager: this.roomManager
    });
  }
  /**
   * Register an auth provider.
   */
  useAuth(provider) {
    this.authManager.register(provider);
    return this;
  }
  /**
   * Register a LiveRoom class.
   * Can be called before start() to register room types dynamically.
   */
  useRoom(roomClass) {
    this.roomRegistry.register(roomClass);
    return this;
  }
  /**
   * Start the LiveServer: register WS + HTTP handlers on the transport.
   */
  async start() {
    if (this.options.componentsPath) {
      await this.registry.autoDiscoverComponents(this.options.componentsPath);
    }
    const wsConfig = {
      path: this.options.wsPath ?? DEFAULT_WS_PATH,
      onOpen: (ws) => this.handleOpen(ws),
      onMessage: (ws, message, isBinary) => this.handleMessage(ws, message, isBinary),
      onClose: (ws, code, reason) => this.handleClose(ws, code, reason),
      onError: (ws, error) => this.handleError(ws, error)
    };
    await this.transport.registerWebSocket(wsConfig);
    if (this.options.httpPrefix !== false) {
      const prefix = this.options.httpPrefix ?? "/api/live";
      await this.transport.registerHttpRoutes(this.buildHttpRoutes(prefix));
    }
    if (this.options.cluster) {
      await this.options.cluster.start();
    }
    if (this.transport.start) {
      await this.transport.start();
    }
    liveLog("lifecycle", null, `LiveServer started (ws: ${wsConfig.path}${this.options.cluster ? ", cluster: enabled" : ""})`);
  }
  /**
   * Graceful shutdown.
   */
  async shutdown() {
    this.registry.cleanup();
    this.connectionManager.shutdown();
    this.fileUploadManager.shutdown();
    this.stateSignature.shutdown();
    if (this.options.cluster) await this.options.cluster.shutdown();
    if (this.transport.shutdown) await this.transport.shutdown();
    liveLog("lifecycle", null, "LiveServer shut down");
  }
  // ===== WebSocket Handlers =====
  handleOpen(ws) {
    const origin = ws.data?.origin;
    const allowedOrigins = this.options.allowedOrigins;
    if (allowedOrigins && allowedOrigins.length > 0) {
      if (!origin || !allowedOrigins.includes(origin)) {
        liveLog("websocket", null, `Connection rejected: origin '${origin || "none"}' not in allowedOrigins`);
        ws.close(4003, "Origin not allowed");
        return;
      }
    }
    const connectionId = `ws-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
    ws.data = {
      connectionId,
      components: /* @__PURE__ */ new Map(),
      subscriptions: /* @__PURE__ */ new Set(),
      connectedAt: /* @__PURE__ */ new Date(),
      origin
    };
    this.connectionManager.registerConnection(ws, connectionId);
    sendImmediate(ws, JSON.stringify({
      type: "CONNECTION_ESTABLISHED",
      connectionId,
      timestamp: Date.now()
    }));
    liveLog("websocket", null, `Connection opened: ${connectionId}`);
  }
  async handleMessage(ws, rawMessage, isBinary) {
    const connectionId = ws.data?.connectionId;
    if (connectionId) {
      const limiter = this.rateLimiter.get(connectionId);
      if (!limiter.tryConsume()) {
        sendImmediate(ws, JSON.stringify({ type: "ERROR", error: "Rate limit exceeded", timestamp: Date.now() }));
        return;
      }
    }
    if (isBinary && rawMessage instanceof ArrayBuffer) {
      try {
        const { header, data } = decodeBinaryChunk(rawMessage);
        if (header.type === "FILE_UPLOAD_CHUNK") {
          const chunkMessage = { ...header, data: "" };
          const progress = await this.fileUploadManager.receiveChunk(chunkMessage, data);
          if (progress) sendImmediate(ws, JSON.stringify(progress));
        }
      } catch (error) {
        sendImmediate(ws, JSON.stringify({ type: "ERROR", error: error.message, timestamp: Date.now() }));
      }
      return;
    }
    const str = typeof rawMessage === "string" ? rawMessage : new TextDecoder().decode(rawMessage);
    if (str.length > MAX_MESSAGE_SIZE) {
      sendImmediate(ws, JSON.stringify({ type: "ERROR", error: "Message too large", timestamp: Date.now() }));
      return;
    }
    let message;
    try {
      message = JSON.parse(str);
    } catch {
      sendImmediate(ws, JSON.stringify({ type: "ERROR", error: "Invalid JSON", timestamp: Date.now() }));
      return;
    }
    if (message.payload) {
      message.payload = sanitizePayload(message.payload);
    }
    try {
      if (message.type === "AUTH") {
        const authContext = await this.authManager.authenticate(message.payload || {});
        if (ws.data) ws.data.authContext = authContext;
        sendImmediate(ws, JSON.stringify({
          type: "AUTH_RESPONSE",
          success: authContext.authenticated,
          payload: authContext.authenticated ? { userId: authContext.user?.id } : { error: "Authentication failed" },
          timestamp: Date.now()
        }));
        return;
      }
      if (message.type === "ROOM_JOIN" || message.type === "ROOM_LEAVE" || message.type === "ROOM_EMIT" || message.type === "ROOM_STATE_SET" || message.type === "ROOM_STATE_GET") {
        await this.handleRoomMessage(ws, message);
        return;
      }
      if (message.type === "FILE_UPLOAD_START") {
        const result2 = await this.fileUploadManager.startUpload(message, ws.data?.userId);
        sendImmediate(ws, JSON.stringify({
          type: "FILE_UPLOAD_START_RESPONSE",
          componentId: message.componentId,
          uploadId: message.payload?.uploadId,
          success: result2.success,
          error: result2.error,
          requestId: message.requestId,
          timestamp: Date.now()
        }));
        return;
      }
      if (message.type === "FILE_UPLOAD_CHUNK") {
        const progress = await this.fileUploadManager.receiveChunk(message);
        if (progress) sendImmediate(ws, JSON.stringify(progress));
        return;
      }
      if (message.type === "FILE_UPLOAD_COMPLETE") {
        const result2 = await this.fileUploadManager.completeUpload(message);
        sendImmediate(ws, JSON.stringify(result2));
        return;
      }
      if (message.type === "COMPONENT_REHYDRATE") {
        const result2 = await this.registry.rehydrateComponent(
          message.componentId,
          message.payload.component,
          message.payload.signedState,
          ws,
          { room: message.payload.room, userId: message.userId }
        );
        sendImmediate(ws, JSON.stringify({
          type: "COMPONENT_REHYDRATED",
          componentId: message.componentId,
          success: result2.success,
          result: result2.success ? { newComponentId: result2.newComponentId } : void 0,
          error: result2.error,
          requestId: message.requestId,
          timestamp: Date.now()
        }));
        return;
      }
      const result = await this.registry.handleMessage(ws, message);
      if (result !== null) {
        const response = {
          type: message.type === "CALL_ACTION" ? "ACTION_RESPONSE" : "MESSAGE_RESPONSE",
          originalType: message.type,
          componentId: message.componentId,
          success: result.success,
          result: result.result,
          error: result.error,
          requestId: message.requestId,
          responseId: message.responseId,
          timestamp: Date.now()
        };
        sendImmediate(ws, JSON.stringify(response));
      }
    } catch (error) {
      sendImmediate(ws, JSON.stringify({
        type: "ERROR",
        componentId: message.componentId,
        error: error.message,
        requestId: message.requestId,
        timestamp: Date.now()
      }));
    }
  }
  handleClose(ws, code, reason) {
    const connectionId = ws.data?.connectionId;
    const componentCount = ws.data?.components?.size || 0;
    this.registry.cleanupConnection(ws);
    this.roomManager.cleanupComponent(connectionId || "");
    if (connectionId) {
      this.connectionManager.cleanupConnection(connectionId);
      this.rateLimiter.remove(connectionId);
    }
    liveLog("websocket", null, `Connection closed: ${connectionId} (${componentCount} components)`);
  }
  handleError(ws, error) {
    console.error(`[LiveServer] WebSocket error:`, error.message);
  }
  // ===== Room Message Router =====
  async handleRoomMessage(ws, message) {
    const { componentId } = message;
    const roomId = message.roomId || message.payload?.roomId;
    switch (message.type) {
      case "ROOM_JOIN": {
        if (this.roomRegistry.resolveFromId(roomId)) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Room requires server-side join via component action",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        const connRooms = ws.data?.rooms;
        if (connRooms && connRooms.size >= MAX_ROOMS_PER_CONNECTION) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Room limit exceeded",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        if (this.authManager.hasProviders()) {
          const authContext = ws.data?.authContext;
          const authResult = await this.authManager.authorizeRoom(
            authContext || ANONYMOUS_CONTEXT,
            roomId
          );
          if (!authResult.allowed) {
            sendImmediate(ws, JSON.stringify({
              type: "ERROR",
              componentId,
              error: authResult.reason || "Room access denied",
              requestId: message.requestId,
              timestamp: Date.now()
            }));
            break;
          }
        }
        const result = this.roomManager.joinRoom(componentId, roomId, ws, message.payload?.initialState);
        if ("rejected" in result && result.rejected) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: result.reason,
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        if (!ws.data.rooms) ws.data.rooms = /* @__PURE__ */ new Set();
        ws.data.rooms.add(roomId);
        sendImmediate(ws, JSON.stringify({
          type: "ROOM_JOINED",
          componentId,
          payload: { roomId, state: result.state },
          requestId: message.requestId,
          timestamp: Date.now()
        }));
        break;
      }
      case "ROOM_LEAVE":
        this.roomManager.leaveRoom(componentId, roomId);
        ws.data?.rooms?.delete(roomId);
        sendImmediate(ws, JSON.stringify({
          type: "ROOM_LEFT",
          componentId,
          payload: { roomId },
          requestId: message.requestId,
          timestamp: Date.now()
        }));
        break;
      case "ROOM_EMIT": {
        if (!this.roomManager.isInRoom(componentId, roomId)) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Not a member of this room",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        this.roomManager.emitToRoom(roomId, message.payload?.event, message.payload?.data, componentId);
        break;
      }
      case "ROOM_STATE_SET": {
        if (!this.roomManager.isInRoom(componentId, roomId)) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Not a member of this room",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        if (this.roomManager.isServerOnlyState(roomId)) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Room state is server-only",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        this.roomManager.setRoomState(roomId, message.payload?.state, componentId);
        break;
      }
      case "ROOM_STATE_GET": {
        if (!this.roomManager.isInRoom(componentId, roomId)) {
          sendImmediate(ws, JSON.stringify({
            type: "ERROR",
            componentId,
            error: "Not a member of this room",
            requestId: message.requestId,
            timestamp: Date.now()
          }));
          break;
        }
        const state = this.roomManager.getRoomState(roomId);
        sendImmediate(ws, JSON.stringify({
          type: "ROOM_STATE",
          componentId,
          payload: { roomId, state },
          requestId: message.requestId,
          timestamp: Date.now()
        }));
        break;
      }
    }
  }
  // ===== HTTP Monitoring Routes =====
  buildHttpRoutes(prefix) {
    return [
      {
        method: "GET",
        path: `${prefix}/stats`,
        handler: () => ({
          body: {
            components: this.registry.getStats(),
            rooms: this.roomManager.getStats(),
            connections: this.connectionManager.getSystemStats(),
            uploads: this.fileUploadManager.getStats(),
            performance: this.performanceMonitor.getStats()
          }
        }),
        metadata: { summary: "Live Components system statistics", tags: ["live"] }
      },
      {
        method: "GET",
        path: `${prefix}/components`,
        handler: () => ({
          body: { names: this.registry.getRegisteredComponentNames() }
        }),
        metadata: { summary: "List registered component names", tags: ["live"] }
      },
      {
        method: "POST",
        path: `${prefix}/rooms/:roomId/messages`,
        handler: (req) => {
          const roomId = req.params.roomId;
          this.roomManager.emitToRoom(roomId, "message:new", req.body);
          return { body: { success: true, roomId } };
        },
        metadata: { summary: "Send message to room via HTTP", tags: ["live", "rooms"] }
      },
      {
        method: "POST",
        path: `${prefix}/rooms/:roomId/emit`,
        handler: (req) => {
          const roomId = req.params.roomId;
          const { event, data } = req.body;
          this.roomManager.emitToRoom(roomId, event, data);
          return { body: { success: true, roomId, event } };
        },
        metadata: { summary: "Emit custom event to room via HTTP", tags: ["live", "rooms"] }
      }
    ];
  }
};

// src/rooms/RoomStateManager.ts
function createTypedRoomState() {
  const rooms = /* @__PURE__ */ new Map();
  const getKey = (type, roomId) => `${type}:${roomId}`;
  return {
    get(type, roomId, defaultState) {
      const key = getKey(type, roomId);
      const room = rooms.get(key);
      if (room) return room.state;
      rooms.set(key, { state: defaultState, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
      return defaultState;
    },
    update(type, roomId, updates) {
      const key = getKey(type, roomId);
      const room = rooms.get(key);
      if (room) {
        room.state = { ...room.state, ...updates };
        room.lastUpdate = Date.now();
        return room.state;
      }
      const newState = updates;
      rooms.set(key, { state: newState, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
      return newState;
    },
    set(type, roomId, state) {
      const key = getKey(type, roomId);
      const room = rooms.get(key);
      if (room) {
        room.state = state;
        room.lastUpdate = Date.now();
      } else {
        rooms.set(key, { state, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
      }
    },
    join(type, roomId) {
      const room = rooms.get(getKey(type, roomId));
      if (room) room.componentCount++;
    },
    leave(type, roomId) {
      const key = getKey(type, roomId);
      const room = rooms.get(key);
      if (room) {
        room.componentCount--;
        if (room.componentCount <= 0) {
          setTimeout(() => {
            const current = rooms.get(key);
            if (current && current.componentCount <= 0) {
              rooms.delete(key);
            }
          }, 5 * 60 * 1e3);
        }
      }
    },
    has(type, roomId) {
      return rooms.has(getKey(type, roomId));
    },
    delete(type, roomId) {
      return rooms.delete(getKey(type, roomId));
    },
    getStats() {
      const roomStats = {};
      for (const [key, info] of rooms) {
        roomStats[key] = { componentCount: info.componentCount, stateKeys: Object.keys(info.state) };
      }
      return { totalRooms: rooms.size, rooms: roomStats };
    }
  };
}
var RoomStateManager = class {
  rooms = /* @__PURE__ */ new Map();
  get(roomId, defaultState) {
    const room = this.rooms.get(roomId);
    if (room) return room.state;
    if (defaultState) {
      this.rooms.set(roomId, { state: defaultState, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
      return defaultState;
    }
    return {};
  }
  update(roomId, updates) {
    const room = this.rooms.get(roomId);
    if (room) {
      room.state = { ...room.state, ...updates };
      room.lastUpdate = Date.now();
      return room.state;
    }
    const newState = updates;
    this.rooms.set(roomId, { state: newState, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
    return newState;
  }
  set(roomId, state) {
    const room = this.rooms.get(roomId);
    if (room) {
      room.state = state;
      room.lastUpdate = Date.now();
    } else {
      this.rooms.set(roomId, { state, componentCount: 0, createdAt: Date.now(), lastUpdate: Date.now() });
    }
  }
  join(roomId) {
    const room = this.rooms.get(roomId);
    if (room) room.componentCount++;
  }
  leave(roomId) {
    const room = this.rooms.get(roomId);
    if (room) {
      room.componentCount--;
      if (room.componentCount <= 0) {
        setTimeout(() => {
          const current = this.rooms.get(roomId);
          if (current && current.componentCount <= 0) {
            this.rooms.delete(roomId);
          }
        }, 5 * 60 * 1e3);
      }
    }
  }
  has(roomId) {
    return this.rooms.has(roomId);
  }
  delete(roomId) {
    return this.rooms.delete(roomId);
  }
  getStats() {
    const rooms = {};
    for (const [roomId, info] of this.rooms) {
      rooms[roomId] = { componentCount: info.componentCount, stateKeys: Object.keys(info.state) };
    }
    return { totalRooms: this.rooms.size, rooms };
  }
};

// src/rooms/InMemoryRoomAdapter.ts
var InMemoryRoomAdapter = class {
  rooms = /* @__PURE__ */ new Map();
  // ===== IRoomStorageAdapter =====
  async getOrCreateRoom(roomId, initialState) {
    const existing = this.rooms.get(roomId);
    if (existing) {
      return { state: existing.state, created: false };
    }
    const now = Date.now();
    const data = {
      state: initialState ?? {},
      createdAt: now,
      lastUpdate: now
    };
    this.rooms.set(roomId, data);
    return { state: data.state, created: true };
  }
  async getState(roomId) {
    return this.rooms.get(roomId)?.state ?? {};
  }
  async updateState(roomId, updates) {
    const room = this.rooms.get(roomId);
    if (room) {
      Object.assign(room.state, updates);
      room.lastUpdate = Date.now();
    }
  }
  async hasRoom(roomId) {
    return this.rooms.has(roomId);
  }
  async deleteRoom(roomId) {
    return this.rooms.delete(roomId);
  }
  async getStats() {
    const rooms = {};
    for (const [id, data] of this.rooms) {
      rooms[id] = {
        createdAt: data.createdAt,
        lastUpdate: data.lastUpdate,
        stateKeys: Object.keys(data.state)
      };
    }
    return { totalRooms: this.rooms.size, rooms };
  }
  // ===== IRoomPubSubAdapter =====
  // No-ops for single-instance: all events are already propagated locally
  // by RoomEventBus and LiveRoomManager's broadcastToRoom().
  async publish(_roomId, _event, _data) {
  }
  async subscribe(_roomId, _handler) {
    return () => {
    };
  }
  async publishMembership(_roomId, _action, _componentId) {
  }
  async publishStateChange(_roomId, _updates) {
  }
};

export { ANONYMOUS_CONTEXT, AnonymousContext, AuthenticatedContext, BINARY_ROOM_EVENT, BINARY_ROOM_STATE, ComponentRegistry, ConnectionRateLimiter, DEFAULT_CHUNK_SIZE, DEFAULT_WS_PATH, EMIT_OVERRIDE_KEY, FileUploadManager, InMemoryRoomAdapter, LiveAuthManager, LiveComponent, LiveRoom, LiveRoomManager, LiveServer, PROTOCOL_VERSION, PerformanceMonitor, RateLimiterRegistry, RoomEventBus, RoomRegistry, RoomStateManager, StateSignatureManager, WebSocketConnectionManager, buildRoomFrame, buildRoomFrameTail, createTypedRoomEventBus, createTypedRoomState, decodeBinaryChunk, encodeBinaryChunk, getLiveComponentContext, jsonCodec, liveLog, liveWarn, msgpackCodec, parseRoomFrame, prependMemberHeader, queueWsMessage, registerComponentLogging, resolveCodec, sendBinaryImmediate, sendImmediate, setLiveComponentContext, unregisterComponentLogging };
